<!DOCTYPE html>
<!-- 
Copyright 2014 EUBrazilCC (EU‐Brazil Cloud Connect)

Licensed under the EUPL, Version 1.1 or - as soon they will be approved by 
the European Commission - subsequent versions of the EUPL (the "Licence");
You may not use this work except in compliance with the Licence.
You may obtain a copy of the Licence at:

  http://ec.europa.eu/idabc/eupl

Unless required by applicable law or agreed to in writing, software 
distributed under the Licence is distributed on an "AS IS" basis,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the Licence for the specific language governing permissions and 
limitations under the Licence.

This product combines work with different licenses. See the "NOTICE" text
file for details on the various modules and licenses.
The "NOTICE" text file is part of the distribution. Any derivative works
that you distribute must include a readable copy of the "NOTICE" text file.
-->
<html lang="en">
<head>
<meta name="description" content="Leishmaniasis Virtual Laboratory (LVL) documentation">
<meta name="keywords" content="Leishmaniasis,LVL,EUBrazilCC">
<meta name="author" content="EUBrazilCC consortium">
<meta charset="UTF-8">
<title>LVL - Administration &amp; Development Guide</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css" media="screen">
<link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootswatch/3.3.2/readable/bootstrap.min.css">
</head>
<body>
  <div class="container">

    <div class="page-header">
      <div class="row">
        <div class="col-xs-12">
          <h1>Leishmaniasis Virtual Laboratory (LVL) Administration &amp; Development Guide</h1>
          <p class="lead">Version 0.0.1 - February 2015</p>
          <p>
            <strong>Abstract.</strong> This document is a guide to the administration and configuration of the Leishmaniasis Virtual Laboratory (LVL).
          </p>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">Index</h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">
            <ul>
              <li><a href="#about">About this guide</a></li>
              <li><a href="#installation">Installation</a>
                <ul>
                  <li><a href="#installation_requirements">Requirements</a></li>
                  <li><a href="#installation_quick_guide">Quick install guide</a></li>
                  <li><a href="#installation_from_source">Installing from source</a></li>
                  <li><a href="#installation_virtual_appliances">Deploying EGI virtual appliances</a></li>
                  <li><a href="#installation_troubleshooting">Troubleshooting</a></li>
                </ul></li>               
              <li><a href="#configuration">Configuration</a>
                <ul>
                  <li><a href="#configuration_common">Common configuration</a></li>
                  <li><a href="#configuration_service">Configuring the LVL service</a></li>
                  <li><a href="#configuration_idp_authx">Configuring the LVL IdP &amp; AuthX server</a></li>
                </ul></li>                
              <li><a href="#security">Security</a>
                <ul>
                  <li><a href="#security_general">General considerations</a></li>
                  <li><a href="#security_external_services_and_apps">External services and applications</a></li>
                  <li><a href="#security_oauth2_authorization">OAuth 2.0 authorization</a></li>
                </ul></li>
              <li><a href="#database">Database</a>
                <ul>
                  <li><a href="#database_configuration">Configuration</a></li>
                  <li><a href="#database_compatibility_with_genBank_and_pubmed">Compatibility with GenBank and PubMed</a></li>
                </ul></li>
              <li><a href="#restful_api">RESTful API</a>
                <ul>
                  <li><a href="#restful_api_resources">Resources</a></li>
                  <li><a href="#restful_api_endpoints">Endpoints</a></li>
                  <li><a href="#restful_api_json_support">JSON support</a>
                    <ul>
                      <li><a href="#restful_api_geojson_support">GeoJSON support</a></li>
                    </ul></li>
                  <li><a href="#restful_api_search">Search</a>
                    <ul>
                      <li><a href="#restful_api_search_sequences">Search sequences</a></li>
                      <li><a href="#restful_api_e-compendium">Leishmaniasis e-compendium</a></li>
                    </ul></li>
                </ul></li>
              <li><a href="#user_interface">User Interface</a>
                <ul>
                  <li><a href="#user_interface_run_locally">Run locally</a></li>
                  <li><a href="#user_interface_programming_models">Programming models</a></li>
                  <li><a href="#user_interface_movile_versions">Mobile versions</a></li>
                  <li><a href="#user_interface_resources">UI resources</a>
                    <ul>
                      <li><a href="#user_interface_country_flags">Country flags</a></li>
                    </ul></li>
                </ul></li>
              <li><a href="#query_language">Query language</a>
                <ul>
                  <li><a href="#query_language_definition">LVL query language definition</a></li>                  
                </ul></li>
              <li><a href="#examples">Examples</a>
                <ul>
                  <li><a href="#examples_restful_api">RESTful API examples</a>
                    <ul>
                      <li><a href="#examples_restful_api_submit_sequence">Submit a new sequence</a></li>
                      <li><a href="#examples_restful_api_get_sequence">Get a sequence using the accession number</a></li>
                      <li><a href="#examples_restful_api_update_sequence">Update a sequence</a></li>
                      <li><a href="#examples_restful_api_delete_sequence">Delete a sequence</a></li>
                      <li><a href="#examples_restful_api_list_sequences">List sequences</a></li>
                      <li><a href="#examples_restful_api_find_nearby_sequences">Find nearby sequences</a></li>
                      <li><a href="#examples_restful_api_import_sequences">Import new sequences into the database</a></li>
                      <li><a href="#examples_restful_api_export_public_link_to_sequence">Export a public link to a sequence</a></li>
                    </ul></li>
                  <li><a href="#examples_oauth2">OAuth2 examples</a>
                    <ul>
                      <li><a href="#examples_oauth2_get_access_token_password">Get access token for user/password</a></li>
                      <li><a href="#examples_oauth2_access_to_protected_resource">Access to protected resource</a></li>
                    </ul></li>
                  <li><a href="#examples_tasks">Task examples</a>
                    <ul>
                      <li><a href="#examples_tasks_import_sequences">Import sequences from external databases (PubMed)</a></li>
                    </ul></li>
                </ul></li>
              <li><a href="#limitations">Known limitations</a>
                <ul>
                  <li><a href="#limitations_usage_limits">Usage limits</a></li>
                  <li><a href="#limitations_desired_features">Desired features list</a></li>
                </ul></li>
            </ul>
          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="about"></a>About this guide
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <p>
              This guide uses the Bootstrap theme <a href="http://bootswatch.com/readable/" target="_blank">Readable</a> from WrapBootstrap, which is optimized
              for legibility.
            </p>

          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="installation"></a>Installation
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">
            <h2>
              <a id="installation_requirements"></a>Requirements
            </h2>

            <p>
              In order to install a complete LVL system, you need at least one server running Linux (Ubuntu, Debian, CentOS). Currently, the recommended
              platform to run LVL is <a href="http://www.ubuntu.com/download/server" target="_blank">Ubuntu 14.04 LTS</a>. It is highly recommended to use a
              separate server instance (virtual or physical) for the LVL collection to obtain better performance.
            </p>

            <p>A minimum LVL installation includes the following components:</p>

            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading"><a id="lvl_lvl-collection"></a>LVL collection</h4>
                <p class="list-group-item-text">
                  A database based on <a href="https://www.mongodb.org/" target="_blank">MongoDB</a>. <em>Recommended version:</em> 2.6.7. Also, a message 
                  broker based in <a href="http://activemq.apache.org/" target="_blank">Apache ActiveMQ</a> is provided to coordinate the operation among 
                  the <i>LVL service</i> replicas. <em>Recommended version:</em> 5.6.0 (embedded version 5.10.1).
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading"><a id="lvl_lvl-authx"></a>LVL identity provider &amp; authorization server</h4>
                <p class="list-group-item-text">
                  A Java servlet application that runs in a servlet container. <em>Recommended setup:</em> <a href="http://tomcat.apache.org/" target="_blank">Apache
                    Tomcat</a> 7.0.52.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading"><a id="lvl_lvl-service"></a>LVL service</h4>
                <p class="list-group-item-text">
                  A servlet-based REST service. <em>Recommended setup:</em> it can be combined with the <i>LVL identity provider &amp; authorization server</i>
                  in the same servlet container or it can run separately in a different container.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading"><a id="lvl_lvl-portal"></a>LVL Web portal</h4>
                <p class="list-group-item-text">
                  A pure-JavaScript application that needs an HTTP server. <em>Recommended setup:</em> <a href="http://nginx.org/" target="_blank">nginx</a>
                  1.4.6. Optionally, you can include <a href="https://github.com/nbs-system/naxsi/wiki" target="_blank">Naxsi</a> with nginx to protect your
                  server against <a href="http://en.wikipedia.org/wiki/Cross-site_scripting" target="_blank">XSS</a> &amp; 
                  <a href="http://en.wikipedia.org/wiki/SQL_injection" target="_blank">SQL injection</a>.
                </p>                
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading"><a id="lvl_workflow-manager"></a>Workflow manager</h4>
                <p class="list-group-item-text">
                  Workflow execution requires a workflow engine that is responsible for interpreting the workflow specification, and for scheduling and
                  coordinating the different tasks. In EUBrazilCC this component is based on <a href="http://www.esciencecentral.co.uk/" target="_blank">e-Science Central</a>, 
                  which is a platform for data analysis. <em>Recommended version:</em> 3.1-SNAPSHOT.
                </p>                
              </li>
            </ul>
            
            <div class="alert alert-warning">
              <div class="media">
                <div class="pull-left">
                  <h1>
                    <span class="glyphicon glyphicon glyphicon-exclamation-sign"></span>
                  </h1>
                </div>
                <div class="media-body">
                  <h4 class="media-heading">Public links</h4>
                  <p>
                    Besides supporting the execution of the user interface, the HTTP server can also serve to provide access to the data that is anonymously 
                    shared through public links. It's highly recommended that this data can be accessed outside the RESTful API to avoid overloading the
                    Java application server. When the option <code>portal.public</code> is available in the LVL 
                    <a href="https://github.com/eubrazilcc/leishmaniasis-virtual-lab/blob/master/lvl-project/lvl-resource-bundle/src/main/etc/lvl.xml" target="_blank" class="alert-link">main configuration</a>
                    a link will be automatically injected in the public link under the tag <code>downloadLink</code>. See the 
                    <a href="#examples_restful_api_export_public_link_to_sequence" class="alert-link">examples</a> bellow for further details.
                  </p>
                </div>
              </div>
            </div>

            <p>For example, the following command will install all these required dependencies in Ubuntu:</p>

            <code> $ sudo apt-get install openjdk-7-jre tomcat7 nginx mongodb activemq </code>
            <p></p>

            <p>Or, including Naxsi:</p>

            <code> $ sudo apt-get install openjdk-7-jre tomcat7 nginx-naxsi mongodb activemq </code>
            <p></p>

            <p>
              However, for a better performance, it is recommended to follow the instructions bellow to install and <a href="#database_configuration">configure
              the database</a> with the latest MongoDB production version.
            </p>

            <h4>Testing email</h4>

            <p>A mailing system is required to send notifications. The following example sends an email from the command line:</p>

            <code> $ echo "This is the message body" | mail -s "This is the subject" user@example.com -aFrom:Name\&lt;from_user@example.com\&gt; </code>
            <p></p>

            <h2>
              <a id="installation_quick_guide"></a>Quick install guide
            </h2>

            <p>This is a simple guide to get LVL setup and running quickly. For additional details on installation and configuration you can read the rest
              of this guide.</p>

            <h3>Setup directories</h3>

            <p>Create the LVL base directory and extract the resource bundle to the new directory:</p>

            <code> $ sudo mkdir /opt/lvl </code>

            <br />

            <code> $ sudo tar xvzf lvl-resource-bundle-0.0.1.tar.gz -C /opt/lvl </code>

            <p></p>

            <p>
              The LVL configuration files are located in the directory <em>/opt/lvl/etc</em>, you can customize these files to your needs.
            </p>

            <p>
              Finally, change the owner of the <em>/opt/lvl/</em> to the Tomcat user:
            </p>

            <code> $ sudo chown -R tomcat7.tomcat7 /opt/lvl </code>
            <p></p>

            <h3>Setup database</h3>

            <p>
              Currently, LVL only supports password-based authentication for MongoDB connections. If you enable this <a
                href="http://docs.mongodb.org/manual/security/" target="_blank">security mechanism in MongoDB</a>, then you will need to update the LVL
              configuration file: <em>/opt/lvl/etc/lvl.xml</em>.
            </p>
            
            <h3>Setup message broker</h3>
            
            <p>
              An Apache ActiveMQ installation is embedded within the LVL distribution. Both the <i><a href="#lvl_lvl-service">LVL service</a></i> 
              and the <i><a href="#lvl_lvl-authx">LVL identity provider &amp; authorization server</a></i> can be configured to start a              
              message broker with the LVL. This broker uses the transporter <code>nio://localhost:61616</code> by default and only one
              instance should be started per <code>JVM</code>. The brokers instantiated in this way are interconnected in a network of
              brokers.
            </p>
            
            <h4>Setting and external message broker</h4>
            
            <p>
              Recommended configuration for Apache ActiveMQ is based on the <a href="http://activemq.apache.org/xml-configuration.html" target="_blank">default configuration</a>.
              In particular, it is recommended to configure NIO <a href="http://activemq.apache.org/configuring-transports.html" target="_blank">transport </a> 
              with the <a href="http://activemq.apache.org/tcp-transport-reference.html" target="_blank">available options</a>. Additional 
              configuration options can be applied to protect your message broken from DOS attacks, such as limiting concurrent connections 
              and frame size.
            </p>
            
            <p>
              The <code>systemUsage</code> controls the maximum amount of space the broker will use before slowing down <a href="http://activemq.apache.org/producer-flow-control.html" target="_blank">producers</a>.
              When using ActiveMQ in a small virtual machine (1 core, 2GB RAM, 20GB disk storage) - the following limits could safely be used:
            </p>
            
            <ul>
              <li>Memory Usage: amount of JVM memory that is available for the broker. <i>Recommended: </i> 128MB.</li>
              <li>Store Usage: the disk space used by persistent messages (using KahaDB). <i>Recommended: </i> 5GB.</li>
              <li>Temp Usage: the disk space used by non-persistent message, assuming we are using default KahaDB. ActiveMQ spools non-persistent 
                  messages to disk in order to prevent broker running out of memory. <i>Recommended: </i> 1GB.</li>
            </ul>
            
            <p>
              <strong>FIXIT: </strong> Additional configuration parameters can be set in the file <code>/etc/default/activemq</code>. For example, 
              include the following line to force Apache ActiveMQ to use a thread pool to dispatch messages:
            </p>
            
            <code>ACTIVEMQ_OPTS="$ACTIVEMQ_OPTS -Dorg.apache.activemq.UseDedicatedTaskRunner=false"</code>
            <p></p>
            
            <p>
              Additional recommendations to achieve vertical and horizontal scaling can be found in the following
              <a href="http://www.javabeat.net/deploying-activemq-for-large-numbers-of-concurrent-applications/" target="_blank">link</a>.
              For a general description about the supported topologies and configurations read the Apache ActiveMQ
              <a href="http://activemq.apache.org/networks-of-brokers.html" target="_blank">online documentation</a>.              
            </p>
            
            <p>
              Notes to the Ubuntu installation: logs are stored by default in the file: <code>/var/lib/activemq/main/data/activemq.log</code>.
            </p>
            
            <div class="alert alert-success">
              <div class="media">
                <div class="pull-left">
                  <h1>
                    <span class="glyphicon glyphicon glyphicon-ok-sign"></span>
                  </h1>
                </div>
                <div class="media-body">
                  <h4 class="media-heading">Apache ActiveMQ configuration example</h4>
                  <p>
                    The following configuration file: <a href="config/activemq.xml.txt" target="_blank" class="alert-link">activemq.xml</a>
                    provides an example for installing an external Apache ActiveMQ service that can be safely used with the LVL.
                  </p>
                </div>
              </div>
            </div>
            
            <h3>Install &amp; configure services</h3>

            <p>
              (In Ubuntu, by default <em>CATALINA_HOME</em> is <em>/usr/share/tomcat7</em>, and <em>CATALINA_BASE</em> is <em>/var/lib/tomcat7</em>. Both
              variables can be set in <em>/etc/default/tomcat7</em>)
            </p>

            <p>Copy or link the common LVL dependencies to the Tomcat shared libraries directory. These dependencies could include logging and database
              drivers, which are known to perform better when loaded only once within the same JVM. For example:</p>

            <code> $ ln -s /opt/lvl/lib/java/mongo-java-driver-2.12.0.jar /usr/share/tomcat7/lib/mongo-java-driver-2.12.0.jar </code>
            <p></p>

            <p>The following script is an example of how this task can be done in an automatic manner:</p>

            <code> for f in `ls /opt/lvl/lib/java/*.jar`; do ln -s ${f} /usr/share/tomcat7/lib/`basename ${f}`; done </code>
            <p></p>

            <p>Copy the LVL services to the Tomcat Web applications directory and change the owner of the files to the Tomcat user:</p>

            <code> $ sudo cp lvl-auth.war lvl-service.war /var/lib/tomcat7/webapps/ </code>

            <br />

            <code> $ sudo chown -R tomcat7.tomcat7 /var/lib/tomcat7/webapps/*.war </code>
            <p></p>

            <p>Service deployment should occur automatically and you can track the progress in the Tomcat log file:</p>

            <code> $ sudo tail -f /var/log/tomcat7/catalina.out </code>
            <p></p>

            <p>LVL services will produce their own log files:</p>

            <code>
              $ sudo tail -f /opt/lvl/var/log/lvl/lvl-auth.log <br /> $ sudo tail -f /opt/lvl/var/log/lvl/lvl-service.log
            </code>
            <p></p>

            <h3>Install &amp; configure Web portal</h3>
            
            <p>Configure your HTTP server as usual: assign an HTTP port (usually 80), a root directory for your installation and a server name. Copy the
              content of the <code>htdocs</code> directory of the LVL distribution to your root directory. Optionally, enable compression on your server 
              and tune-up other parameters, such as expiration time, etc.</p>
              
            <p>The LVL requires two services to be enabled in the HTTP server, which are served by a Java Application Server. You can simply configure
              your HTTP server to proxy your Java services. For example, in nginx, add two location aliases in the <code>server</code> section of the 
              configuration to proxy your Java services:</p>
              
<pre>
location ^~ /lvl-auth/ {
    proxy_pass http://127.0.0.1:8080;
    proxy_redirect   off;
    proxy_set_header Host            $host;
    proxy_set_header X-Real-IP       $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_read_timeout 5m;
    proxy_connect_timeout 120;
}
</pre>
            
<pre>
location ^~ /lvl-service/ {
    proxy_pass http://127.0.0.1:8080;
    proxy_redirect   off;
    proxy_set_header Host            $host;
    proxy_set_header X-Real-IP       $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_read_timeout 5m;
    proxy_connect_timeout 120;
}
</pre>
            
            <p>Create an alias to your folder where the public files are stored. For example, in nginx, add a location alias in the <code>server</code>
              section of the configuration to serve the public files from the filesystem in a static manner:</p>
            
<pre>
location /lvl-pub/ {
    alias /opt/lvl/data/shared/;
}
</pre>

            <h4>Securing your connection</h4>

            <p>If you are going to provide access to the LVL using a secure chennel (<strong>*RECOMMENDED*</strong>) you will need to configure your
              HTTP server to support HTTPS connections. For example, in nginx, add the following to the <code>server</code> section of the configuration:</p>
              
<pre>
listen 443 ssl;

keepalive_timeout 70;

ssl_certificate /etc/nginx/ssl/cert.pem;
ssl_certificate_key /etc/nginx/ssl/key.pem;

ssl_session_cache   shared:SSL:10m;
ssl_session_timeout 10m;

ssl_protocols SSLv3 TLSv1 TLSv1.1 TLSv1.2;
ssl_ciphers RC4:HIGH:!aNULL:!MD5;
ssl_prefer_server_ciphers on;
</pre>

            <p>Export your server certificate and private key:</p>

<pre>
openssl pkcs12 -in lvl_i3m_upv_es@pkirisgrid.p12 -nocerts -nodes -out key.pem
openssl pkcs12 -in lvl_i3m_upv_es@pkirisgrid.p12 -nokeys -nodes -out cert.pem
</pre>

            <p>Configure your certificate permissionsand restart your HTTP server:</p>

<pre>
root@lvl:~# chown root.www-data /etc/nginx/ssl/*.pem
root@lvl:~# chmod 0640 /etc/nginx/ssl/*.pem
root@lvl:~# ls -lh /etc/nginx/ssl/
total 8.0K
-rw-r----- 1 root www-data 1.9K Dec 17 13:57 cert.pem
-rw-r----- 1 root www-data 1.9K Dec 17 13:57 key.pem
</pre>

            <p>More details in the nginx online <a href="http://nginx.org/en/docs/http/configuring_https_servers.html" target="_blank">documentation</a>.</p>

            <h4>Access to the sandbox endpoint</h4>
            
            <p>Install the <a href="http://pki.irisgrid.es/ca/cert/" target="_blank">pkIRISGrid CA</a> certificate in your browser to trust the server 
               certificate presented by the sandbox. This is the PKI CA for the spanish e-science, which is part of the European Grid Infrastructure (EGI).</p>           

            <h3>Performance optimization</h3>

            <p>
              In order to improve performance, you can use a load balancer (nginx) to distribute the workload among several replicas of the services. This
              option is currently available for the <em>LVL identity provider &amp; authorization server</em> and the <em>LVL service</em>. Database performance
              can be improved by deploying more replicas of the <em>LVL collection</em>.
            </p>

            <h2>
              <a id="installation_from_source"></a>Installing from source
            </h2>

            <p>
              To build the LVL system you will need the Java Development Kit (JDK). Currently, the recommended Java implementation to build LVL is <a
                href="http://openjdk.java.net/projects/jdk7u/" target="_blank">OpenJDK 7</a>. Note that JDK is only needed for building LVL. For just running
              LVL, a JRE (Java Runtime Environment) will suffice.
            </p>

            <p>
              In addition to JDK, you will also need <a href="http://maven.apache.org/" target="_blank">Apache Maven</a> and a <a href="http://git-scm.com/"
                target="_blank">Git</a> client.
            </p>

            <p>Clone the LVL source code repository:</p>

            <code> $ git clone https://github.com/eubrazilcc/leishmaniasis-virtual-lab.git </code>
            <p></p>

            <p>Enter to the LVL project main folder and install the LVL project BOM and POM files in your local Maven repository:</p>

            <code> $ cd leishmaniasis-virtual-lab </code>
            <br />
            <code> $ mvn install:install-file -Dfile=pom.xml -DpomFile=pom.xml </code>
            <br />
            <code> $ mvn clean install -pl lvl-project </code>
            <p></p>

            <p>Enter to the project folder, build and install the LVL core libraries in your local Maven repository:</p>

            <code> $ cd lvl-project </code>
            <br />
            <code> $ mvn clean install -pl lvl-core,lvl-storage </code>            
            <p></p>

            <p>Build the LVL services:</p>

            <code> $ mvn clean package -pl lvl-auth </code>            
            <br />
            <code> $ mvn clean package -pl lvl-service </code>
            <p></p>

            <p>
              Copy the LVL service packages to your server and continues with the usual <a href="#installation">installation</a> steps:
            </p>

            <code> $ scp lvl-auth/target/lvl-auth.war lvl-service/target/lvl-service.war
              &lt;USERNAME&gt;@&lt;SERVER_HOSTNAME&gt;:&lt;SERVER_LOCATION&gt;/ </code>
            <p></p>

            <p>Build the LVL resources bundle and copy it to your server:</p>

            <code> $ mvn clean package -pl lvl-resource-bundle </code>

            <br />

            <code> $ scp lvl-resource-bundle/target/lvl-resource-bundle-0.0.1.tar.gz &lt;USERNAME&gt;@&lt;SERVER_HOSTNAME&gt;:&lt;SERVER_LOCATION&gt;/ </code>
            <p></p>

            <h3>Integration tests</h3>

            <p>
              Currently, there are several accepted approaches as to how to handle Web service integration test with Maven. The LVL project uses the <em>verify</em>
              goal of <em>the Maven Failsafe Plugin</em>, which is one of the solutions suggested by a member of the Apache Maven Project Management Committee
              (PMC) in <a href="http://stackoverflow.com/a/16936585" target="_blank">response</a> to a stackoverflow subscriber question.
            </p>

            <p>In order to build (including JUnit tests) a LVL service, you should run the common Maven command:</p>

            <code> $ mvn clean package -pl lvl-auth </code>
            <p></p>

            <p>However, the previous process does not include integration tests. In order to force the build process to pass through this type of test you
              should build LVL services with the following Maven command:</p>

            <code> $ mvn clean verify -pl lvl-auth </code>
            <p></p>

            <p>
              This will start an Apache Tomcat 7 container after the package is created, it will deploy the service WAR (<em>lvl-auth.war</em> in this example)
              to the container and will perform additional tests that resembles user interaction in the production environment. If the service fails to pass the
              integration test, the build process will be aborted by Maven.
            </p>

            <p>Currently, the LVL packages feature integration tests that include the following:</p>

            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading">lvl-storage</h4>
                <p class="list-group-item-text">Database deployment in MongoDB 2.6.x</p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">lvl-auth</h4>
                <p class="list-group-item-text">Database deployment in MongoDB 2.6.x, and Web service deployment in Apache Tomcat 7.</p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">lvl-service</h4>
                <p class="list-group-item-text">Database deployment in MongoDB 2.6.x, and Web service deployment in Apache Tomcat 7.</p>
              </li>
            </ul>

            <h3>Skipping JUnit tests</h3>

            <p>
              To skip running the tests for the entire project or for a particular project, set the <em>maven.test.skip</em> property to <em>true</em> in the
              command line to skip test compilation and execution. See the question <a href="http://maven.apache.org/general.html#skip-test" target="_blank">How
                do I skip the tests?</a> in the Maven's documentation for more details.
            </p>

            <p>
              For example, to build the <em>LVL RESTful Web service</em> with all its dependencies, skipping test you can execute:
            </p>

            <code> $ mvn clean install -pl lvl-service -am -Dmaven.test.skip=true </code>
            <p></p>
            
            <h2>
              <a id="installation_virtual_appliances"></a>Deploying EGI virtual appliances
            </h2>
            
            <p>
              The preferred manner to install the LVL is to use the LVL  virtual appliances stored in the EGI Applications database (<a href="http://appdb.egi.eu/" target="_blank">EGI AppDB</a>).
              These appliances are pre-configured and tested virtual machine images, ready to run in the EUBrazilCC infrastructure. The following contextualization
              options are available:
            </p>
            
            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading">LVL App front-end</h4>
                <p class="list-group-item-text">
                  Deploys a <i><a href="#lvl_lvl-portal">LVL Web portal</a></i>, providing the LVL user interface and the broker. Requires an IP 
                  address accessible to your expected users. Normally, only one instance is needed.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">LVL App service and data store</h4>
                <p class="list-group-item-text">
                  Deploys a <i><a href="#lvl_lvl-collection">LVL collection</a></i>, providing additional data storage space. This is combined with 
                  a <i><a href="#lvl_lvl-service">LVL service</a></i> and a <i><a href="#lvl_lvl-authx">LVL identity provider &amp; authorization server</a></i>,
                  which increase the LVL capacity to handle concurrent requests. Requires the IP address of your <i><a href="#lvl_lvl-portal">LVL Web portal</a></i>.
                </p>
              </li>                        
            </ul>
            
            <p>
              In addition, the following roles of e-SC are also required:
            </p>            
            
            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading">e-SC service</h4>
                <p class="list-group-item-text">
                  Deploys a <i><a href="#lvl_workflow-manager">Workflow manager</a></i>. Requires the IP address of your <i><a href="#lvl_lvl-portal">LVL Web portal</a></i>.
                  Normally, only one instance is needed.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">e-SC working node</h4>
                <p class="list-group-item-text">
                  Deploys a working node that can be used to execute workflow tasks, increasing the LVL capacity to execute concurrent requests. 
                  Requires the IP address of your <i><a href="#lvl_workflow-manager">Workflow manager</a></i>.
                </p>
              </li>                       
            </ul>
            
            <h2>
              <a id="installation_troubleshooting"></a>Troubleshooting
            </h2>

            <h3>Out of memory error</h3>

            <p>Increase the size of your heap if you get the following error in Tomcat:</p>

            <code>
              SEVERE: Error waiting for multi-thread deployment of WAR files to complete<br /> java.util.concurrent.ExecutionException:
              java.lang.OutOfMemoryError: Java heap space
            </code>
            <p></p>

            <p>
              In Ubuntu, heap size can be increased by modifying the value assigned to the Java startup parameter <em>-Xmx1024m</em> in the <em>JAVA_OPTS</em>.
              This variable is setup in <em>/etc/default/tomcat7</em>.
            </p>

            <p>Restart Tomcat to apply the change:</p>

            <code> $ sudo service tomcat7 restart </code>
            <p></p>

            <h3>
              No <em>Access-Control-Allow-Origin</em> header is present on the requested resource
            </h3>

            <p>This error can be caused by client or server misconfiguration. Open a connection to your server and check that the following headers are
              exchanged between your client and server endpoints:</p>

            <ul>
              <li>Your client request must include a valid accept header:
                <ul>
                  <li>Accept: */*</li>
                </ul>
              </li>
              <li>Your client request must include the origin and the server response header must allow the origin you sent in your request:
                <ul>
                  <li>Origin: http://localhost</li>
                  <li>Access-Control-Allow-Origin: http://localhost</li>
                </ul>
              </li>
              <li>Your client request must include the requested method and the server response header must allow the requested method:
                <ul>
                  <li>Access-Control-Request-Method: POST</li>
                  <li>Access-Control-Allow-Methods: POST</li>
                </ul>
              </li>
              <li>Your client must include the <em>requested with</em> header:
                <ul>
                  <li>Access-Control-Request-Headers: X-Requested-With</li>
                </ul>
              </li>
            </ul>

            <p>The following example shows how an AJAX client connects to the server before requesting a resource:</p>

            <code> $ curl -H "Origin: http://localhost" -H "Access-Control-Request-Method: POST" \ -H "Access-Control-Request-Headers: X-Requested-With"
              -X OPTIONS --verbose \ http://172.16.166.135:8080/lvl-auth/oauth2/v1/pending_users </code>
            <p></p>

            <pre>
* About to connect() to 172.16.166.135 port 8080 (#0)
*   Trying 172.16.166.135... connected
&gt; OPTIONS /lvl-auth/oauth2/v1/pending_users HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: 172.16.166.135:8080
&gt; Accept: */*
&gt; Origin: http://localhost
&gt; Access-Control-Request-Method: POST
&gt; Access-Control-Request-Headers: X-Requested-With
&gt; 
&lt; HTTP/1.1 200 OK
&lt; Server: Apache-Coyote/1.1
&lt; Access-Control-Allow-Origin: http://localhost
&lt; Access-Control-Allow-Credentials: true
&lt; Access-Control-Max-Age: 86400
&lt; Access-Control-Allow-Methods: POST
&lt; Access-Control-Allow-Headers: content-type,access-control-request-headers,accept,access-control-request-method,origin,authorization,x-requested-with
&lt; Content-Length: 0
&lt; Date: Sun, 20 Apr 2014 14:21:10 GMT
&lt; 
* Connection #0 to host 172.16.166.135 left intact
* Closing connection #0
</pre>

            <p>A 200 response means that the client can continue with the operation. If request is invalid, or is not permitted, then request is rejected
              with HTTP status code 403 (Forbidden).</p>
              
            <h3>Compression is not enabled</h3>
            
            <p>By default, GZIP compression is enabled selectively in the LVL REST API for those methods that returns a very large response to the 
              caller. This compression is performed at the level of the RESTful web service as described here, and includes the complete body of the
              response entity sent to the caller:</p>
              
            <ul>
              <li><a href="https://jersey.java.net/documentation/latest/user-guide.html#d0e8359" target="_blank">Jersey User Guide - Interceptors</a></li>
              <li><a href="http://www.javacodegeeks.com/2014/11/how-to-compress-responses-in-java-rest-api-with-gzip-and-jersey.html" target="_blank">How to compress responses in Java REST API with GZip and Jersey</a></li>
            </ul>
            
            <p>While this approach has the advantage of performing the compression in the server where the LVL service runs, the request headers are
              not available when the compression is performed, so the response will be compressed ALWAYS, regardless of the <code>Accept-encoding</code>
              sent by the caller.</p>
            
            <p>As an alternative to the REST API compression, you can optionally configure your HTTP server to compress the content it serves from the 
              LVL services. There are two places where you can configure compression:</p>
              
            <p>The standard HTTP connectors provided by Apache Tomcat supports the configuration attribute <code>compression</code>, however the
              configuration is tricky to work as it depends on the sendfile feature to be enabled or not in the connector as described here: 
              <a href="http://tomcat.apache.org/tomcat-7.0-doc/config/http.html" target="_blank">Apache Tomcat 7, The HTTP Connector: Standard Implementation</a></p>
              
            <p>The second option (recomended) is to enable compression on the load balancer. For example, in nginx, add a GZIP compression block in 
              the server section of the configuration to compress CSS style sheets, text files and JSON-encoded documents:</p>
            
            <pre>
gzip on;              
gzip_types text/css text/plain application/xml application/json application/x-javascript text/javascript;
gzip_comp_level 5;
# gzip_proxied no-store no-cache private expired auth;
gzip_proxied any;
</pre>

            <p>The following example shows the responde header sent by an HTTP server where GZIP compression is enabled:</p>
            
            <code> $ curl -I -H 'Authorization: Bearer oeZseY/z/3T/LG8R9o4AjKI+J7ySQUcoWF0N7YVL7Kc=' -H 'Accept-encoding: gzip,deflate' 
              'http://lvl.i3m.upv.es/lvl-service/rest/v1/sandflies?per_page=1' </code>
            <p></p>

            <pre>
HTTP/1.1 200 OK
Server: nginx/1.4.6 (Ubuntu)
Date: Mon, 15 Dec 2014 15:34:40 GMT
Content-Type: application/json
Connection: keep-alive
Content-Encoding: gzip
</pre>

            <p>The same server will sent the following response header when GZIP compression is disabled:</p>

            <pre>
HTTP/1.1 200 OK
Server: nginx/1.4.6 (Ubuntu)
Date: Mon, 15 Dec 2014 16:00:15 GMT
Content-Type: application/json
Content-Length: 809
Connection: keep-alive
</pre>

            <p>The following example shows the responde body sent by an HTTP server where GZIP compression is enabled. Note that the response is
              decompressed (<code>gunzip</code>) and formatted (Python <code>json.tool</code>) before showing it in the console:</p>

            <code> $ curl -s -H 'Authorization: Bearer oeZseY/z/3T/LG8R9o4AjKI+J7ySQUcoWF0N7YVL7Kc=' -H 'Accept-encoding: gzip,deflate' 
              'http://lvl.i3m.upv.es/lvl-service/rest/v1/sandflies?per_page=1' | gunzip | python -m json.tool</code>
              
<p></p>

            <pre>
{
    "elements": [
        {
            "accession": "AB174769",
            "collected": null,
            "countryFeature": "Ecuador",
            "dataSource": "GenBank",
            "definition": "Lutzomyia ayacuchensis gene for 18S rRNA, partial sequence",
            "gene": [],
            "gi": 52546157,
            "id": "gb:AB174769",
            "length": 405,
            "links": [
                {
                    "href": "/lvl-service/rest/v1/sandflies/gb:AB174769",
                    "params": {
                        "rel": "self",
                        "type": "application/json"
                    }
                }
            ],
            "locale": "_EC",
            "location": {
                "coordinates": [
                    -78.18340599999999,
                    -1.831239
                ],
                "type": "Point"
            },
            "organism": "Lutzomyia ayacuchensis",
            "pmids": [
                "15728872"
            ],
            "version": "AB174769.1"
        }
    ],
    "links": [
        {
            "href": "/lvl-service/rest/v1/sandflies?page=1&amp;per_page=1&amp;q=&amp;sort=&amp;order=asc",
            "params": {
                "rel": "next",
                "type": "application/json"
            }
        },
        {
            "href": "/lvl-service/rest/v1/sandflies?page=22125&amp;per_page=1&amp;q=&amp;sort=&amp;order=asc",
            "params": {
                "rel": "last",
                "type": "application/json"
            }
        }
    ],
    "totalCount": 22126
}
</pre>

          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="configuration"></a>Configuration
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <p>
              The LVL configuration is splitted across the following three files in order to facilitate configuration:
            </p>
            
            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading">Common configuration</h4>
                <p class="list-group-item-text">
                  The file <code>lvl.xml</code> contains common configuration directives that are globally available to all the components
                  of the LVL.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">LVL service configuration</h4>
                <p class="list-group-item-text">
                  The file <code>lvl-service.xml</code> contains configuration directives that are applied only to the LVL service.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">LVL IdP &amp; AuthX server configuration</h4>
                <p class="list-group-item-text">
                  The file <code>lvl-auth.xml</code> contains common configuration directives that applied only to the LVL identity
                  provider and authorization server.
                </p>
              </li>
            </ul>
            
            <p>
              The LVL find the available configuration files, searching the different sources in the following order:
            </p>
             
            <ol>
              <li>Application environment: if the environment variable <code>LVL_HOME</code> is setup, then the files are loaded from
                  the location specified in the variable;</li>
              <li>Default location <code>/opt/lvl</code> in the file-system locally available to the application; and finally</li>
	          <li>If none of the above works, configuration files are returned from the resources available in the Java class path.</li>
	        </ol>
            
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="configuration_common"></a>Common configuration
            </h2>
            
            <ul>
              <li>
                <code>lvl-root</code>: The top of the directory tree under which the service's configuration, supporting libraries 
                and applications are kept. <span class="label label-default">Default</span> <code>/opt/lvl</code>
              </li>
              <li>
                <code>storage.local-cache</code>: The performance of certain applications may be improved by using a local cache (ensure 
                that this directory is writable by the user that LVL runs as). Temporary directory is preferable when your application 
		        uses sensitive data. <span class="label label-default">Default</span> <code>{system temporary directory}/.lvl/cache</code>
              </li>            
              <li>
                <code>database.name</code>: The name of the database where the LVL collections are stored. 
                <span class="label label-default">Default</span> <code>lvldb</code>
              </li>            
              <li>
                <code>database.credentials</code>: Optional user identity to access the database. <span class="label label-default">Default</span>
                If undefined, anonymous access is performed.
              </li>            
              <li>
                <code>database.hosts</code>: Optional list of database servers, specified with the format: <code>hostname:port</code>.
                <span class="label label-default">Default</span> <code>localhost:27017</code>
              </li>
              <li>
                <code>messaging.broker.embedded</code>: Deploy a message broker within the LVL. Otherwise, an external broker should 
                be provided. <span class="label label-default">Default</span> <code>true</code>
              </li>
              <li>
                <code>messaging.hosts</code>: Optional list of message brokers, specified with the format: <code>hostname:port</code>.
                <span class="label label-default">Default</span> <code>localhost:61616</code>
              </li>
              <li>
                <code>smtp.host</code>: Optional SMTP host. <span class="label label-default">Default</span> <code>localhost</code>
              </li>
              <li>
                <code>smtp.port</code>: Optional SMTP port. <span class="label label-default">Default</span> <code>25</code>
              </li>
              <li>
                <code>smtp.support-email</code>: Support email address. <span class="label label-warning">No default</span>
              </li>
              <li>
                <code>smtp.noreply-email</code>: Difussion email addres. <span class="label label-warning">No default</span>
              </li>
              <li>
                <code>workflow.endpoint</code>: Endpoint where the workflow manager is listening for incoming requests.
                <span class="label label-warning">No default</span>
              </li>
              <li>
                <code>workflow.credentials</code>: User identity to access the workflow manager.
                <span class="label label-warning">No default</span>
              </li>
            </ul>
            
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="configuration_service"></a>Configuring the LVL service
            </h2>
            
            <ul>
              <li>Prefix: <code>rest-service</code></li>
            </ul>
            
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="configuration_idp_authx"></a>Configuring the LVL IdP &amp; AuthX server
            </h2>
            
            <ul>
              <li>Prefix: <code>authz-server</code></li>
            </ul>
            
          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="security"></a>Security
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="security_general"></a>General considerations
            </h2>

            <p>At least, the identification provider and authorization server should run on HTTPS, although it's also recommended to use HTTPS with the
              RESTful API. The following is a general description of how to create server keys &amp; certificates in Java to test your deployment with HTTPS.</p>

            <p>Generate server key:</p>

            <code> $ keytool -genkey -keystore ./server-keystore -alias serverKey -dname "CN=localhost, OU=GRyCAP, O=I3M-UPVLC, L=Valencia, ST=Spain,
              C=ES" -validity 360 </code>
            <p></p>

            <p>Export server certificate:</p>

            <code> $ keytool -export -alias serverKey -rfc -keystore ./server-keystore &gt; ./server.cert </code>
            <p></p>

            <p>Import server certificate to client trust-store:</p>

            <code> $ keytool -import -alias serverCert -file ./server.cert -keystore ./client-truststore </code>
            <p></p>
            
            <h2>
            	<a id="security_external_services_and_apps"></a>External services and applications
            </h2>
            
            <h3>Configuring LVL to access e-Science Central (e-SC) over HTTPS</h3>
            
            <p>The e-SC certificate must be importd to the Java trust store before starting the LVL service both to execute JUnit and interation
              tests as well as to deploy a LVL server in production.</p>
            
            <p>Valid e-SC certificates are provided in the <a href="http://eubrazilcc-rm.i3m.upv.es/projects/eubrazilcc/wiki/WP3_infrastructure">Wiki</a>
              of the project EUBrazilCC.</p>
            
            <h4>Installing e-SC certificate in production</h4>
            
            <p>First, download the e-SC service certificate from the EUBrazilCC site:</p>
            
            <code>$ wget http://eubrazilcc-rm.i3m.upv.es/attachments/download/22/upv-vm-2.crt</code>
            <p></p>
            
            <p>It's highly recommended to check the validity of the certificate before installing it to the Java trusted repository:</p>
            
            <code>$ openssl x509 -in upv-vm-2.crt -noout -text</code>
            <p></p>

<pre>
...

Issuer: C=ES, ST=Catalonia, L=Valencia, O=Newcastle University, OU=School of Computing Science, CN=158.42.105.23
Validity
    Not Before: Oct  1 23:00:54 2014 GMT
    Not After : Dec 30 23:00:54 2014 GMT
Subject: C=ES, ST=Catalonia, L=Valencia, O=Newcastle University, OU=School of Computing Science, CN=158.42.105.23
</pre>

            <p>Install the certificate in the default Java key store:</p>
            
            <code>$ sudo keytool -import -alias esc-server -file upv-vm-2.crt -keystore $JAVA_HOME/jre/lib/security/cacerts</code>
            <p></p>

<pre>
Enter keystore password:  (Default Password: changeit)

...

Trust this certificate? [no]:  yes

Certificate was added to keystore
</pre>

            <p>Check that the certificate is accessible from Java:</p>

            <code>$ keytool -list -alias esc-server -keystore $JAVA_HOME/jre/lib/security/cacerts</code>
            <p></p>

<pre>
esc-server, Oct 10, 2014, trustedCertEntry, 
Certificate fingerprint (SHA1): 90:C9:9B:7E:0E:64:65:0C:E0:08:E2:A2:8F:E5:20:20:10:BF:5C:93
</pre>

            <p>Optionally, you can create a link in your local system to configure Java to use your key store as the default key store:</p>

            <code>$ ln -s $JAVA_HOME/jre/lib/security/cacerts $HOME/.keystore</code>
            <p></p>

            <p>Now, you don't have to specify the path to the Java key store:</p>

            <code>$ keytool -list -alias esc-server</code>
            <p></p>

<pre>
esc-server, Oct 10, 2014, trustedCertEntry, 
Certificate fingerprint (SHA1): 90:C9:9B:7E:0E:64:65:0C:E0:08:E2:A2:8F:E5:20:20:10:BF:5C:93
</pre>

            <p>To remove the certificate from the Java key store:</p>

            <code>$ keytool -delete -alias esc-server -keystore $JAVA_HOME/jre/lib/security/cacerts</code>
            <p></p>

            <h4>Installing e-SC certificate in development or testbed</h4>            
            
            <code> $ keytool -import -alias esc-server -file ./esc-server.cer -keystore ./client-truststore </code>
            <p></p>

            <h2>
              <a id="security_oauth2_authorization"></a>OAuth 2.0 authorization
            </h2>
            
            <p>LVL clients must include an HTTP authorization request header as defined in the 
              <a href="https://tools.ietf.org/html/rfc6750#section-2.1" target="_blank">section 2.1</a> 
              - "Authorization Request Header Field" of the RFC 6750:</p>
            
<pre>Authorization: Bearer CelWb/k6GP3heYhrPBSa4wZGSSdW9HM1TuGZrOyE/+0=</pre>

            <p>Alternatively, a query parameter can be used to send the access token in the HTTP request URI as defined in the 
              <a href="https://tools.ietf.org/html/rfc6750#section-2.3" target="_blank">section 2.3</a> - "URI Query Parameter" of the RFC 6750:</p>

<pre>GET /resource?access_token=CelWb/k6GP3heYhrPBSa4wZGSSdW9HM1TuGZrOyE/+0= HTTP/1.1</pre>

            <div class="alert alert-success">
              <div class="media">
                <div class="pull-left">
                  <h1>
                    <span class="glyphicon glyphicon glyphicon-exclamation-sign"></span>
                  </h1>
                </div>
                <div class="media-body">
                  <h4 class="media-heading">LVL authenticated requests</h4>
                  <p>
                    The preferred method to send the bearer access tokens in resource requests to the LVL server is the <code>Authorization</code> 
                    request header field. Clients <strong>MUST NOT</strong> use more than one method to transmit the token in each request.
                  </p>
                </div>
              </div>
            </div>

            <p>
              <em>TODO:</em> improve this section with a better description.
            </p>

            <h3>Roles</h3>

            <ul>
              <li><em>Resource owner:</em> Any registered user with an active LVL account can grant access to individual resources or collections in the
                LVL.
                <ul>
                  <li>All LVL users.</li>
                </ul></li>
              <li><em>Identity provider (IdP):</em> User identity consists of: username, e-mail address and password.
                <ul>
                  <li>LVL identity service (integrated in the authorization service in the current version).
                    <ul>
                      <li>Each resource identified by a URI has an assigned scope that coincides with the base name of the resource to facilitate access
                        control. For example, the resource <i>/sequences</i> will use the rules defined for the scope <i>sequences</i> to grant access.
                      </li>
                      <li>Scopes can be annotated with additional attributes, which can be later used to grant access to the resource. By default, read
                        only access is granted to the resources in a scope.</li>
                      <li>To enable a user with unlimited full-access, add him/her to the scope annotated with the attribute <em>a</em>. For example: <i>sequences,a</i>.
                      </li>
                      <li>Multiple attributes separated by comma can be assigned to the scope.</li>
                      <li>In the current version, identity is provided along with the authorization request (OAuth 2.0 <a
                        href="http://tools.ietf.org/html/rfc6749#section-4.3" target="_blank">Resource Owner Password Credentials Grant</a> authorization use
                        case). Other expected use case is the OAuth 2.0 <a href="http://tools.ietf.org/html/rfc6749#section-4.1" target="_blank">Authorization
                          Code Grant</a>, where registered clients will be granted with access to the <em>default</em> scope.
                      </li>
                    </ul>
                  </li>
                </ul></li>
              <li><em>Client application:</em> Any end-user application that access the resources (information, computing and storage capacities) from a
                user interface on behalf of an authenticated LVL user.
                <ul>
                  <li>LVL Web portal.</li>
                </ul></li>
              <li><em>Resource server:</em> API providers.
                <ul>
                  <li>LVL RESTful Web service.</li>
                </ul></li>
              <li><em>Authorization &amp; token server:</em> The server issuing access tokens to the client after successfully authenticating the resource
                owner and obtaining authorization.
                <ul>
                  <li>LVL authorization service (implemented with <a href="http://oltu.apache.org/" target="_blank">Apache Oltu</a>).
                  </li>
                </ul></li>
            </ul>

            <h3>Scope aliases</h3>

            <ul>
              <li><em>all</em>: grant global access to all LVL resources
                <ul>
                  <li>users/*,a</li>
                  <li>sequences,a</li>
                  <li>pending-sequences,a</li>
                  <li>pipelines,a</li>
                  <li>my-sequences/*,a</li>
                  <li>public_links,a</li>
                </ul></li>
              <li><em>user</em>: grant read-only access to public LVL resources, full-access to user workspace
                <ul>
                  <li>users/&lt;username&gt;,a</li>
                  <li>sequences</li>
                  <li>pipelines</li>
                  <li>my-sequences/&lt;username&gt;,a</li>
                  <li>public_links/&lt;username&gt;,a</li>
                </ul></li>
              <li><em>data-curator</em>: grant full-access to sequences
                <ul>
                  <li>sequences,a</li>
                  <li>pending-sequences,a</li>
                </ul></li>
              <li><em>default</em>: grant read-only access to public LVL resources
                <ul>
                  <li>sequences</li>
                  <li>pipelines</li>
                </ul></li>
            </ul>

            <h3>Authorization schema</h3>

            <p>
              a) LVL Web portal uses the OAuth 2.0 <a href="http://tools.ietf.org/html/rfc6749#section-4.3" target="_blank">Resource Owner Password
                Credentials Grant</a> schema:
            </p>

            <pre>
        username/password
1) User ----------------&gt; Authz Server --&gt; IdP --&gt; Database (user + scopes)
        &lt;----------------
          access token
         scope (optional)
       
       
          access token             check()
2) User ----------------&gt; Resource -------&gt; Database (user + scopes)
</pre>

            <p>
              b) Other (anonymous) clients can use the OAuth 2.0 <a href="http://tools.ietf.org/html/rfc6749#section-4.1" target="_blank">Authorization Code
                Grant</a> schema:
            </p>

            <p>
              <em>TODO</em>
            </p>

            <h3>Resources</h3>

            <ul>
              <li><a href="http://tools.ietf.org/html/rfc6749">RFC6749 - The OAuth 2.0 Authorization Framework</a>.</li>
              <li><a href="https://developers.google.com/oauthplayground/">OAuth 2.0 Playground - test OAuth interactions with Google OAuth2 API</a>.</li>
              <li><a href="http://oltu.apache.org/">Apache Oltu - OAuth protocol implementation in Java</a>.</li>
            </ul>
          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="database"></a>Database
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="database_configuration"></a>Configuration
            </h2>

            <p>
              Follow the instructions from the MongoDB documentation to install a server: <a href="http://docs.mongodb.org/manual/installation/" target="_blank">Install
                MongoDB</a>.
            </p>

            <p>
              In this case we setup a minimal MongoDB installation on Ubuntu: <a href="http://docs.mongodb.org/manual/tutorial/install-mongodb-on-ubuntu/"
                target="_blank">Install MongoDB on Ubuntu</a>.
            </p>

            <h2>
              <a id="database_compatibility_with_genBank_and_pubmed"></a>Compatibility with GenBank and PubMed
            </h2>

            <p>The LVL database resembles other open-access nucleotide databases, in particular GenBank. Indeed, GenBank references are included in the LVL
              database. Sequences are fetched from the Nucleotide database in GBSeqXML format. There are future plans to fetch TinySeqXML sequences to check for
              updates. This format offers less information than GenBank flat files or GBSeqXML, but the size of TinySeqXML records is 40% less than GenBank flat
              files and 60% less than GBSeqXML.</p>

            <p>GBSeqXML format is specified in DTD the following files:</p>

            <code>
              $ wget http://www.ncbi.nlm.nih.gov/dtd/NCBI_GBSeq.dtd <br /> $ wget http://www.ncbi.nlm.nih.gov/dtd/NCBI_Entity.mod.dtd <br /> $ wget
              http://www.ncbi.nlm.nih.gov/dtd/NCBI_GBSeq.mod.dtd
            </code>
            <p></p>

            <p>JAXB Binding Compiler (xjc) can be used to generate the Java binding classes (output directory should exist before calling xjc):</p>

            <code>
              $ xjc -dtd -d output -p eu.eubrazilcc.lvl.core.xml.ncbi.gb -mark-generated NCBI_Entity.mod.dtd <br /> $ xjc -dtd -d output -p
              eu.eubrazilcc.lvl.core.xml.ncbi.gb -mark-generated NCBI_GBSeq.mod.dtd <br /> $ xjc -dtd -d output -p eu.eubrazilcc.lvl.core.xml.ncbi.gb
              -mark-generated NCBI_GBSeq.dtd
            </code>
            <p></p>

            <p>
              Generated files are copied to the project <em>lvl-core</em>.
            </p>

            <p>
              XML binding classes are generated in the same manner for Entrez <em>ESearch</em> utility application:
            </p>

            <code>
              $ wget http://www.ncbi.nlm.nih.gov/corehtml/query/DTD/eSearch_020511.dtd <br /> $ xjc -dtd -d output -p eu.eubrazilcc.lvl.core.xml.ncbi.esearch
              -mark-generated eSearch_020511.dtd
            </code>
            <p></p>
            
            <p>In a similar way, LVL publication references are compatible with PubMed:</p>            
            
            <code>
              $ wget http://www.ncbi.nlm.nih.gov/corehtml/query/DTD/pubmed_100101.dtd <br />
              $ wget http://www.ncbi.nlm.nih.gov/corehtml/query/DTD/nlmmedlinecitationset_100101.dtd <br />
              $ xjc -dtd -d output -p eu.eubrazilcc.lvl.core.xml.ncbi.pubmed -mark-generated pubmed_100101.dtd
            </code>
            <p></p>
            
          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="restful_api"></a>RESTful API
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <p>
              The <em>LVL service</em> was designed to decople the client and the server as much as possible. To this end, the server only provides a RESTful
              JSON API to feed the possible clients, such as the <em>LVL Web portal</em>. As a consequence, all the templating, views and business logic should
              be done client-side. The <em>LVL Web portal</em> is a pure-JavaScript, single-page Web application that resides as a static project in an HTTP
              server, which is completely independent of the <em>LVL API</em> container (the client is not sent from the <em>LVL API</em> server, as is the
              case, for example, in most PHP applications).
            </p>

            <p>Futhermore, the server and the client can be provided on different servers and domains in order to improve performance.</p>

            <p>
              On the other hand, the resources that the <em>LVL API</em> provides are user specific (e.g. DNA sequences, biomolecular pipelines), so the <em>LVL
                identity provider &amp; authorization server</em> provides an authentication and authorization mechanism based on OAuth2 in the backend service.
            </p>

            <p>
              As the clients and the servers may reside on different domains, <a href="http://www.w3.org/TR/cors/" target="_blank">Cross-origin resource
                sharing (CORS)</a> becomes one of the major concerns of this design. To cope with CORS, the following additional HTTP headers may be sent as part of
              server response:
            </p>

            <code>Access-Control-Allow-Origin: *</code>
            <br>
            <code>Access-Control-Allow-Headers: origin, content-type, accept, authorization</code>
            <br>
            <code>Access-Control-Allow-Credentials: true</code>
            <br>
            <code>Access-Control-Allow-Methods: GET PUT DELETE POST HEAD OPTIONS</code>
            <br>
            <code>Access-Control-Max-Age: 604800</code>

            <p></p>

            <p>
              This <a href="http://enable-cors.org/server.html" target="_blank">link</a> provides a list of resources to add CORS support to different HTTP
              servers and technologies. In particular, this <a href="http://tomcat.apache.org/tomcat-7.0-doc/config/filter.html#CORS_Filter" target="_blank">
                Tomcat filter</a> adds support for CORS in Tomcat (starting from Tomcat version 7.0.41).
            </p>

            <p>
              <a href="https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)" target="_blank">Cross-Site Request Forgery (CSRF)</a> is an attack
              that can compromise end user data and operation or even the entire Web application if the targeted end user is the administrator account. To
              protect the RESTful API from this kind of attack, the server may activates the <a
                href="https://jersey.java.net/documentation/latest/client.html#d0e4689" target="_blank">CsrfProtectionFilter</a> Jersey filter, which adds the <em>X-Requested-By</em>
              header to each state changing request. This filter may be combined with the AngularJS <a
                href="https://code.angularjs.org/1.2.16/docs/api/ng/service/$http" target="_blank">$http service</a>, which provides protection to CSRF. An
              alternative could be to activate the Tomcat <a href="http://tomcat.apache.org/tomcat-7.0-doc/config/filter.html#CSRF_Prevention_Filter"
                target="_blank">CSRF Prevention Filter</a>. However, this filter has very specific and strong requirements that are difficult to meet in client
              applications.
            </p>

            <h2>
              <a id="restful_api_resources"></a>Resources
            </h2>

            <p>List of resources accessible through the RESTful API:</p>
            
            <ul>
              <li>/sequences
                <ul>
                  <li>/leismania/{id}
                    <ul>
                      <li>/export/gb (GenBank file format)
                        <ul>
                          <li>/xml</li>
                          <li>/text</li>
                        </ul></li>
                    </ul>
                  </li>
                  <li>/sandflies/{id}
                    <ul>
                      <li>/export/gb (GenBank file format)
                        <ul>
                          <li>/xml</li>
                          <li>/text</li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                </ul>
              </li>
              <li>/citations/{id}
                <ul>
                  <li>/pubmed
                    <ul>
                      <li>/xml</li>
                    </ul>
                  </li>
                  <li>/fulltext
                    <ul>
                      <li>/pdf</li>
                      <li>/html</li>
                    </ul>
                  </li>
                </ul>
              </li>
              <li>/datasets
                <ul>
                  <li>/objects/{ns}/{id}
                    <ul>
                      <li>/download</li>
                    </ul>
                  </li>
                  <li>/shares/{ns}/{id}/{subject}</li>
                  <li>/open_access/{ns}/{id}</li>
                </ul>
              </li>
              <li>/public
                <ul>
                  <li>/datasets/{id}</li>
                </ul>
              </li>
              <li>/pipelines
                <ul>
                  <li>/definitions</li>
                  <li>/runs</li>
                </ul>
              </li>
              <li>/tasks</li>
              <li>/notifications</li>
            </ul>
            
            <p>
              <em>TODO LIST</em>
            </p>

            <ol>
              <li>Get new publications from PubMed and new sequences from GenBank: the system must periodically inspect these resources for new records on
                Leishmaniasis. Users with administrator role are notified when new records are found.</li>
              <li>Provide assisted data import: user with administrator role imports references from PubMed and GenBank.</li>
              <li>The system must at least assist in identifying the following fields: PubMed identifier, sequence accession number, geographical
                coordinates and/or country/region.</li>
            </ol>

            <h2>
              <a id="restful_api_endpoints"></a>Endpoints
            </h2>

            <ul>
              <li>Sandbox endpoint: <a href="http://lvl.i3m.upv.es/lvl-service" target="_blank">http://lvl.i3m.upv.es/lvl-service</a></li>
              <li>Production endpoint: <em>TODO</em></li>
            </ul>

            <h2>
              <a id="restful_api_json_support"></a>JSON support
            </h2>

            <p>
              JSON processing is built on top of the <a href="http://wiki.fasterxml.com/JacksonHome" target="_blank">Jackson JSON processor</a> to provide
              unified annotations across all layers (database records, REST calls, etc.). This includes data binding capabilities to read and write Java objects
              in a seamless way that does not require application developers to adapt their practices to use JSON with the LVL.
            </p>

            <p>
              Jackson is also used as the JSON provider for the JAX-RS 2.0 web service. Currently, the LVL include this <a
                href="https://jersey.java.net/documentation/latest/media.html#json.jackson" target="_blank">Jersey module</a> that supports JSON binding in
              Jersey via Jackson.
            </p>

            <p>
              JAX-RS contains a very useful representation of the <a href="http://www.iana.org/assignments/link-relations/link-relations.xhtml" target="_blank">Link
                Relation</a> that can be serialised with a <a href="https://jax-rs-spec.java.net/nonav/2.0/apidocs/javax/ws/rs/core/Link.JaxbAdapter.html"
                target="_blank">JAXB adapter</a> into XML and JSON. LVL resources use this representation to provide the necessary links as well as reading and
                writing JAX-RS Link objects. Many JSON processing libraries like <a href="http://wiki.fasterxml.com/JacksonFAQ#JAXB" target="_blank">Jackson
                support JAXB annotations</a>. However, native Jackson annotations have been preferred due to problems to serialize links with the current version
                of JAX-RS. The following fragment of <code>Java</code> code shows an example using JAXB annotations:
            </p>
            
            <pre>
import java.util.List;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import javax.ws.rs.core.Link;
import javax.ws.rs.core.Link.JaxbAdapter;

import org.glassfish.jersey.linking.Binding;
import org.glassfish.jersey.linking.InjectLink;
import org.glassfish.jersey.linking.InjectLinks;

@XmlRootElement
public class Sequence {
    @InjectLinks({
        @InjectLink(value="sequences/{id}", rel="self", bindings={@Binding(name="id", value="${instance.id}")}) <br>
    })
    @XmlElement(name = "link")
    @XmlElementWrapper(name = "links")
    @XmlJavaTypeAdapter(JaxbAdapter.class)
    private List&lt;Link&gt; links; // HATEOAS links
    ...
}</pre>

            <p>One issue to be taken into account when using these JAXB annotations with Jackson is that the element that contains the list of links 
              will be named after the <code>@XmlElement</code> name and not the <code>@XmlElementWrapper</code> as expected. This is a
              <a href="https://github.com/FasterXML/jackson-module-jaxb-annotations/issues/13" target="_blank">documented behavior</a>.</p>

            <p>In addition, Jersey annotations have been used to inject the necessary links in a server response using the 
              <a href="https://jersey.java.net/documentation/latest/user-guide.html#declarative-linking" target="_blank">declarative hyperlinking</a> API.
              This module intercepts GET requests and includes the links before sending the response to the caller. Since the links are injected directly
              in the Java class, these annotations are compatible with whatever serialization technology/annotations are used (JAXB, Jackson).
            </p>

            <p>MongoDB stores data in JSON documents. Instead on using a library for mapping Java objects to/from MongoDB, classes are annotated with
              Jackson and JAXB annotations and serialization/deserialization occurs on the fly using Jackson.</p>

            <h3>
              <a id="restful_api_geojson_support"></a>GeoJSON support
            </h3>

            <p>
              Currently, LVL supports the following GeoJSON geometries: point, line and polygon. For a complete description: <a href="http://geojson.org/"
                target="_blank">GeoJSON -- JSON Geometry and Feature Description</a>.
            </p>

            <h2>
              <a id="restful_api_search"></a>Search
            </h2>

            <p>The Search API is designed to help users to find entries (e.g. sequences, papers, users) in the LVL database. It supports two different
              search strategies: a) keyword matching search; and b) full-text search. Search qualifiers support keyword matching search, unless otherwise
              specified. Regardless the searched collection, the following parameters are available to all searches and they can be passed as query parameters
              as part of the URL:</p>

            <ul class="list-group">
              <li class="list-group-item">
                <h4 class="list-group-item-heading">page</h4>
                <p class="list-group-item-text">
                  The current page. The search API uses a 0-based index (the first page is always 0). <strong>Default: </strong> 0.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">per_page</h4>
                <p class="list-group-item-text">The maximum number of entries returned per page. The default value depends on the collection.</p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">q (query)</h4>
                <p class="list-group-item-text">
                  The search terms. Every search term must include a search qualifier and a search expression, especified in the following way:
                  <code>qualifier:expression</code>
                  . To include multiple search terms separate them with a semicolon:
                  <code>q1:exp1;q2:exp2</code>
                  . Some expressions can include symbols:
                  <code>length:&gt;23</code>
                  . See the search terms supported per collection below.
                </p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">sort</h4>
                <p class="list-group-item-text">The sort field. The possible values as well as the default value depends on the collection. When not
                  specified, items will be sorted in the insertion order.</p>
              </li>
              <li class="list-group-item">
                <h4 class="list-group-item-heading">order</h4>
                <p class="list-group-item-text">
                  The sort order if sort parameter is provided. One of asc or desc. <strong>Default:</strong> asc.
                </p>
              </li>
            </ul>

            <div class="alert alert-warning">
              <div class="media">
                <div class="pull-left">
                  <h1>
                    <span class="glyphicon glyphicon glyphicon-exclamation-sign"></span>
                  </h1>
                </div>
                <div class="media-body">
                  <h4 class="media-heading">UI search operators</h4>
                  <p>
                    The LVL search interface is inspired in <a href="https://support.google.com/websearch/answer/136861" target="_blank" class="alert-link">Google</a>
                    to facilitate its use. Think of it the way you think of performing a search on Google. Just like searching on Google, when you search using
                    an operator, don't add any spaces between the operator and your query. A search for
                    <code>accession:AC238558</code>
                    will work, but
                    <code>accession:&nbsp;AC238558</code>
                    will not.
                  </p>
                </div>
              </div>
            </div>

            <h3>
              <a id="restful_api_search_sequences"></a>Search sequences
            </h3>

            <p>Find sequences via various criteria.</p>

            <code>GET /lvl-service/rest/v1/sequences</code>

            <p></p>

            <p>
              100 sequences are returned per page by default, unless other page size is especified using the parameter:
              <code>per_page</code>
            </p>

            <p>
              The
              <code>q</code>
              search term can contain any combination of the supported sequence search qualifiers:
            </p>

            <ul>
              <li><code>source</code> With this qualifier users can restrict the search to just the sequences from a specific data source (e.g. GenBank).</li>
              <li><code>definition</code> Filter sequences by the definition they have.</li>
              <li><code>accession</code> Search for sequences that have accessions that match a certain identifier.</li>
              <li><code>organism</code> Filter sequences based on the organism.</li>
              <li><code>country</code> Filter sequences by the country indicated in the metadata.</li>
              <li><code>locale</code> Filter sequences by the language indicated in the metadata, using two letter country codes to indicate the language
                (e.g. en for English).</li>
              <li><code>text</code> Full-text search that finds the especified string anywhere within the fields data source, definition, accession,
                organism and country of the sequences stored in the LVL database.</li>
            </ul>

            <h4>Example:</h4>

            <p>
              Search for sequences that contains the string "<em>phlebotomus papatasi</em>" and that are derived from the GenBank database:
            </p>

            <code>GET /lvl-service/rest/v1/sequences?q=text:phlebotomus papatasi;source:genbank&amp;per_page=20</code>

            <p></p>

            <p>URL encoded:</p>

            <code>GET /lvl-service/rest/v1/sequences?q=text%3Aphlebotomus+papatasi%3Bsource%3Agenbank&amp;per_page=20</code>

            <h3>
              <a id="restful_api_e-compendium"></a>Leishmaniasis e-compendium
            </h3>

            <p>The leishmaniasis e-compendium is a compilation of all available peer-reviewed literature on leishmaniasis epidemiology. The first version of 
            the e-compendium was compiled in 2010 by the WHO and the release note is available in the following link:
            <a href="http://www.sciencedirect.com/science/article/pii/S147149221000139X" target="_blank">The leishmaniasis e-compendium: a geo-referenced bibliographic tool</a>
            </p>
            
            <p>The LVL include the most actual version of the e-compendium, which was builded from the following search in Nucleotide database from GeBank:</p>
            
            <code>
            (sandfly[Organism] OR sand fly[Organism] OR Phlebotomus[Organism] OR Sergentomyia[Organism] OR Lutzomyia[Organism]) AND (cytochrome oxidase I[Gene Name] OR cytochrome oxidase 1[Gene Name] OR coi[Gene Name] OR COI[Gene Name] OR co1[Gene Name] OR CO1[Gene Name] OR coxi[Gene Name] OR cox1[Gene Name] OR COXI[Gene Name])
            </code>
            
            <p>In the moment of the compilation, there were 815 DNA sequences that match this query in the database and the terms <i>sandfly</i>,
            <i>sand fly</i> and <i>cytochrome oxidase 1</i> did not produce any match.</p>

            <p>
              <em>TODO</em>
            </p>

          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="user_interface"></a>User Interface
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="user_interface_run_locally"></a>Run locally
            </h2>

            <p>You will need an HTTP server to test user interface locally on you machine. The simplest probably is to use Python's built-in http server. If
              you have Python installed, it should be enough to run this from a command line:</p>

            <code>
              # Python 2.x <br /> $ python -m SimpleHTTPServer <br /> <br /> # Python 3.x <br /> $ python -m http.server
            </code>
            <p></p>

            <pre>
Serving HTTP on 0.0.0.0 port 8000 ...
            </pre>

            <p>This will serve files from the current directory at localhost under port 8000:</p>

            <code> http://localhost:8000/ </code>

            <h2>
              <a id="user_interface_programming_models"></a>Programming models
            </h2>

            <p>The UI supports two different programming models depending on the type of access:</p>

            <ul>
              <li>Asynchronous access: possible changes of the information displayed in the UI are not visible until the user explicitly requests the
                refresh, for example, by clicking on a button.
                <ul>
                  <li>HTTP GET (RESTful API).</li>
                </ul>
              </li>
              <li>Synchronous access (near real-time): the UI displays information that is continuously being updated.
                <ul>
                  <li>Server-sent Events (SSE).</li>
                  <li>WebSockets was also considered but quickly discarded due to a few limitations found (mostly lack of support in some browsers).</li>
                </ul>
              </li>
            </ul>

            <h2>
              <a id="user_interface_movile_versions"></a>Mobile versions
            </h2>

            <p>
              A responsive mobile-first design strategy was used to develop the UI. The <a href="http://getbootstrap.com/" target="_blank">Bootstrap</a>
              front-end framework was used with a flat UI design <a href="http://designmodo.github.io/Flat-UI/" target="_blank">theme</a> to develop the
              responsive web design. A single-page application was developed using the <a href="http://backbonejs.org/" target="_blank">Backbone</a> + <a
                href="http://marionettejs.com/" target="_blank">Marionette</a> MVC framework. The <a href="http://ol3js.org/" target="_blank">OpenLayers</a>
              JavaScript library was used to render and display georeferenced information in maps. All this application is modular, with <a
                href="http://requirejs.org/" target="_blank">RequireJS</a> providing the necessary means for loading CSS styles, JavaScript modules and HTML
              templates on-demand.
            </p>

            <!-- <a href="http://jquerymobile.com/" target="_blank">jQuery Mobile</a> was discarded in a first approach due to known incompatibility issues with those frameworks, in special with AngularJS.  -->

            <h2>
              <a id="user_interface_resources"></a>UI resources
            </h2>

            <p>Description of resources used in the design and develop of the user interface.</p>

            <h3>
              <a id="user_interface_country_flags"></a>Country flags
            </h3>

            <p>
              A <a href="http://goo.gl/glz7wT" target="_blank">shell script</a> was created to facilitate the use of country names and flags. This script
              downloads a list of country names from <a href="http://opengeocode.org/" target="_blank">OpenGeocode</a> (a collection of free public domain
              geolocation databases), creates a CSS image sprite with the icon flags obtained from <a href="http://www.geonames.org/" target="_blank">GeoNames</a>
              (an open geographical database) and a <a href="http://requirejs.org/" target="_blank">RequireJS</a> module that allows JS applications to convert
              from the two-letter country codes (ISO 3166-1 alpha-2) to country names. The output of the script includes the following files: <em>flags.css</em>,
              <em>flags.png</em> and <em>country_names.js</em>.
            </p>

            <p>
              The flag icons are converted and assembled using <a href="http://www.imagemagick.org/" target="_blank">ImageMagick</a> and the result image is
              optimized using <a href="http://pmt.sourceforge.net/pngcrush/" target="_blank">Pngcrush</a>.
            </p>

            <p>GeoNames.org provides one of the most complete sets of country flag icons (which has a flat design). Other datasets of flag icons ready for
              use:</p>

            <ul>
              <li><a href="https://github.com/tellnes/country-flags" target="_blank">Country Flags server in Node.JS</a></li>
              <li><a href="http://www.icondrawer.com/flag-icons.php" target="_blank">IconDrawer flag icons</a></li>
              <li><a href="https://github.com/usrz/bootstrap-languages" target="_blank">Language labels and icons for Bootstrap 3</a></li>
            </ul>

          </div>
        </div>
      </div>
    </div>
    
    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="query_language"></a>Query language
            </h1>
            
            <p>Strictly speaking, the LVL is a data-intensive application that provides access to several collections of data curated by experts. Therefore,
              searching and filtering collections are among the most important features of the LVL. To this end, a query language is provided that extends
              the NCBI GenBank <a href="http://www.ncbi.nlm.nih.gov/nuccore/advanced" target="_blank">advanced search</a> language. The rest of this section
              explains the differences between the two languages, providing examples of the query capabilities of the LVL query language and describing
              the grammar syntax and the key semantics of this language.</p>            
            
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="query_language_definition"></a>LVL query language definition
            </h2>
            
            <p>The LVL query language (LvLQL) was developed with <a href="http://www.antlr.org/" target="_blank">ANTLR v4</a>, which is a
              parser generator for structured text. The lexical rules as well as the syntax of the LvLQL language are available in the 
              source code <a href="https://github.com/eubrazilcc/leishmaniasis-virtual-lab/tree/master/lvl-project/lvl-storage/src/main/antlr4/" target="_blank">repository</a>
              of the project. An editor is available for ANTLR v4 grammars: <a href="http://tunnelvisionlabs.com/products/demo/antlrworks" target="_blank">ANTLRWorks 2</a>,
              which can be used to visualize the file format <code>g4</code>.</p>
              
            <p>Unless you are planning to extend the LvLQL language with your own rules, there is no need for installing ANTLR or any other
              tool related with this framework to use the LVL.</p>
              
            <p>Railroad diagrams (RRD) for LvLQL grammar rules are provided <a href="lvlql">here</a>. These diagrams were created using 
              <a href="https://github.com/bkiers/rrd-antlr4" target="_blank">RRD for ANTLR4</a>.</p>
              
            <p>An advanced query builder is included in the LVL portal that allow end-users to rely on a graphical user interface to create
              LvLQL queries that could be applied to the different collections provided with the LVL. Application developers are the expected 
              users of the LvLQL language, while end-users who connect to the LVL through the portal will use this GUI-based query builder.</p>
            
            <h3>LvLQL language features</h3>
            
            <p>The LvLQL language provides users with the means for creating queries that are executed in the data collections provided
              by the LVL. The motivation for this language is in the NCBI GenBank advanced search, but the syntax of the language was
              largelly improved in order to support more advances features than GenBank does. For example, the following search was created
              using the GenBank query language to search the DNA sequences that were obtained from the gen <i>coi</i> of the organism 
              <i>phlebotomus papatasi</i>:</p>
            
            <code>(phlebotomus papatasi[Organism]) AND coi[Gene Name]</code>
            <p></p>
            
            <p>The same search can be expressed using the LvLQL language:</p>
            
            <code>"phlebotomus papatasi":Organism AND "Gene Name":coi</code>
            <p></p>
            
            <p>The main difference is in the use of the double-quote for delimitating values that include spaces, and the use of the colon
              punctuation symbol to separate field names from values.</p>
            
            <p>Beyond these differences in syntax, the LvLQL language support comparison operators that can be used to filter the results.
              For example, limit the results obtained in the previous queries to the sequences with more than 600 bp:</p>
            
            <code>"phlebotomus papatasi":Organism AND "Gene Name":coi AND Length > 600</code>
            <p></p>
            
            <p>One additional feature of the LvLQL language is that supports <a href="http://www.ebi.ac.uk/Tools/sss/" target="_blank">sequence similarity searching</a>
              operators that allows users to filter the sequence collections by similarity with a target sequence or with a species:</p>
            
            <code>SIMSEQ(gb:JX105037) > 90</code>
            <p></p>
            
            <code>SIMSEQ(species:"Phlebotomus papatasi") > 90</code>
            <p></p>
            
            <p>In the first example, the prefix <code>gb</code> is used to indicate that the sequence identifier <code>JX105037</code> 
              links to a DNA sequence in the GenBank database, while the second example the target sequence will be the consensus
              sequence of the species <code>Phlebotomus papatasi</code>.</p>
            
            <h4>LvLQL query examples:</h4>
            
            <dl>
              <dt><code>Phlebotomus OR sandfly</code></dt>
              <dd>Find all the sequences with the strings <code>Phlebotomus</code> or <code>sandfly</code> present in any field of the
                record.</dd>
              <dt><code>Phlebotomus AND NOT papatasi</code></dt>
              <dd>Find all the sequences with the string <code>Phlebotomus</code> present in any field, and the string <code>sandfly</code>
                <strong>*NOT*</strong> being present.</dd>
              <dt><code>Phlebotomus AND "Gene Name":coi AND Length > 600</code></dt>
              <dd>Find all the sequences with the string <code>Phlebotomus</code> present in any field, the string <code>coi</code>
                in the field <code>Gene Name</code>, and the sequence lengh being larger than 600 bp.</dd>
              <dt><code>SIMSEQ(gb:JX105037) > 90</code></dt>
              <dd>Find all the sequences with more than 90% of similaryty with the sequence <code>JX105037</code> from 
                <code>GenBank</code>.</dd>
              <dt><code>SIMSEQ(species:"Phlebotomus papatasi") > 90</code></dt>
              <dd>Find all the sequences with more than 90% of similaryty with the consensus sequence of the species <code>Phlebotomus 
                papatasi</code>.</dd>
              <dt><code>Organism:"phlebotomus papatasi" AND SIMSEQ(gb:JX105037) > 90</code></dt>
              <dd>Find all the sequences reported for the organism <code>Phlebotomus papatasi</code> that share more than 90% of similaryty 
                with the sequence <code>JX105037</code> from <code>GenBank</code>.</dd>
            </dl>            
            
            <h3>Extending the LvLQL language:</h3>
            
            <p>There are several options that you can use to build ANTLR 4 grammar files. ANTLRWorks 2 allows building the parsers from
              the grammar files, the ANTLR 4 <a href="http://www.antlr.org/api/maven-plugin/latest/index.html" target="_blank">Maven plugin</a>
              allows to build the parsers as part of the project building, and finally, the method described below consists in creating
              the parsers statically and test them.</p> 
            
            <p>This section describes the steps performed to build the LvLQL parsers statically. First, the standalone version of ANTLR 4 
              is installed locally:</p>
            
            <code>$ cd /usr/local/lib ; wget http://www.antlr.org/download/antlr-4.4-complete.jar</code>
            <p></p>
            
            <p>Some configuration is needed before you can start using ANTLR 4:</p>
            
            <code>$ export CLASSPATH="$CLASSPATH:/usr/local/lib/antlr-4.4-complete.jar" ; alias antlr4='java -jar /usr/local/lib/antlr-4.4-complete.jar' ; alias grun='java org.antlr.v4.runtime.misc.TestRig'</code>
            <p></p>            
            
            <p>Compiling the LvLQL grammar:</p>
            
            <code>$ cd /tmp/lvlql</code>
            <p></p>
            
            <code>$ rm -f *.java *.class *.tokens ; antlr4 &lt;WORKSPACE&gt;/leishmaniasis-virtual-lab/lvl-project/lvl-storage/src/main/antlr4/eu/eubrazilcc/lvl/storage/ql/antlr4/*.g4 -o . -listener -no-visitor -encoding UTF-8 ; javac *.java</code>
            <p></p>
              
            <p>Checking your changes after extending the LvLQL grammar:</p>
            
            <code>$ grun LvLQL query -gui</code>
            <p></p>
            
            <p>For example, to run the tests of the <code>query</code> target:</p>
            
            <code>$ grun LvLQL query -gui &lt;WORKSPACE&gt;/leishmaniasis-virtual-lab/lvl-project/lvl-storage/src/test/resources/files/antlr4/query-*.txt</code>
            <p></p>
            
            <p>You can visit the ANTLR4 <a href="https://theantlrguy.atlassian.net/wiki/display/ANTLR4/ANTLR+4+Documentation" target="_blank">online documentation</a> 
              for other possible <code>grun</code> options like: tree and tokens.</p>
            
            <p>Notice that the binaries created in this way <strong>*CANNOT*</strong> be used with the LVL project. Since the LVL is build with Maven, you will 
              need the parsers to be generated with the ANTLR 4 <a href="http://www.antlr.org/api/maven-plugin/latest/index.html" target="_blank">Maven plugin</a>.
              Simply copy your grammar files (*.g4) to the LVL source code directory. Don't forget to include the 
              <a href="http://mvnrepository.com/artifact/org.antlr/antlr4-runtime" target="_blank">ANTLR 4 runtime</a> with your application.</p>
            
          </div>
        </div>
      </div>
    </div>

    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="examples"></a>Examples
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">

            <h2>
              <a id="examples_restful_api"></a>RESTful API examples
            </h2>

            <p>
              This section provides simple examples to illustrate the use of the REST API. The examples use <a href="http://curl.haxx.se/" target="_blank">cURL</a>
              to access the endpoint.
            </p>

            <h3>
              <a id="examples_restful_api_submit_sequence"></a>Submit a new sequence:
            </h3>

            <code> $ curl -v -X POST -H "Content-Type: application/json" -d '{"accession": "456", "version": "1.0", "definition": "example sequence",
              "organism": "unknown"}' http://127.0.0.1:8080/lvl/sequences </code>
            <p></p>

            <pre>
* About to connect() to 127.0.0.1 port 8080 (#0)
*   Trying 127.0.0.1... connected
&gt; POST /lvl/sequences HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: 127.0.0.1:8080
&gt; Accept: */*
&gt; Content-Type: application/json
&gt; Content-Length: 95
&gt; 
* upload completely sent off: 95out of 95 bytes
&lt; HTTP/1.1 201 Created
&lt; Location: http://127.0.0.1:8080/lvl/sequences/456
&lt; Date: Thu, 06 Mar 2014 09:46:20 GMT
&lt; Content-Length: 0
&lt; 
* Connection #0 to host 127.0.0.1 left intact
* Closing connection #0
</pre>

            <h3>
              <a id="examples_restful_api_get_sequence"></a>Get a sequence using the accession number:
            </h3>

            <code> $ curl -v http://127.0.0.1:8080/lvl/sequences/456 </code>
            <p></p>

            <pre>
* About to connect() to 127.0.0.1 port 8080 (#0)
*   Trying 127.0.0.1... connected
&gt; GET /lvl/sequences/456 HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: 127.0.0.1:8080
&gt; Accept: */*
&gt; 
&lt; HTTP/1.1 200 OK
&lt; Content-Type: application/json
&lt; Date: Thu, 06 Mar 2014 09:57:38 GMT
&lt; Content-Length: 88
&lt; 
* Connection #0 to host 127.0.0.1 left intact
* Closing connection #0
{"accession":"456","definition":"example sequence","organism":"unknown","version":"2.0"}
</pre>

            <h3>
              <a id="examples_restful_api_update_sequence"></a>Update a sequence:
            </h3>

            <code> $ curl -v -X PUT -H "Content-Type: application/json" -d '{"accession": "456", "version": "2.0", "definition": "example sequence",
              "organism": "unknown"}' http://127.0.0.1:8080/lvl/sequences/456 </code>
            <p></p>

            <pre>
* About to connect() to 127.0.0.1 port 8080 (#0)
*   Trying 127.0.0.1... connected
&gt; PUT /lvl/sequences/456 HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: 127.0.0.1:8080
&gt; Accept: */*
&gt; Content-Type: application/json
&gt; Content-Length: 95
&gt; 
* upload completely sent off: 95out of 95 bytes
&lt; HTTP/1.1 204 No Content
&lt; Date: Thu, 06 Mar 2014 09:48:45 GMT
&lt; 
* Connection #0 to host 127.0.0.1 left intact
* Closing connection #0
</pre>

            <h3>
              <a id="examples_restful_api_delete_sequence"></a>Delete a sequence:
            </h3>

            <code> $ curl -v -X DELETE http://127.0.0.1:8080/lvl/sequences/456 </code>
            <p></p>

            <pre>
* About to connect() to 127.0.0.1 port 8080 (#0)
*   Trying 127.0.0.1... connected
&gt; DELETE /lvl/sequences/456 HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: 127.0.0.1:8080
&gt; Accept: */*
&gt; 
&lt; HTTP/1.1 204 No Content
&lt; Date: Thu, 06 Mar 2014 09:49:57 GMT
&lt; 
* Connection #0 to host 127.0.0.1 left intact
* Closing connection #0
</pre>

            <h3>
              <a id="examples_restful_api_list_sequences"></a>List all the sequences in the database. This method uses pagination, by default only the first 100
              items are retrieved and the navigation links are included in the response:
            </h3>

            <code> $ curl -v -H 'Authorization: Bearer ubauFeNkTWiVGQWTcfJPUW55nWTIY3c+yu25p/Sm7KU=' 'http://lvl.i3m.upv.es/lvl-service/rest/v1/sequences' </code>
            <p></p>

            <pre>
* Hostname was NOT found in DNS cache
*   Trying 158.42.105.62...
* Connected to lvl.i3m.upv.es (158.42.105.62) port 80 (#0)
&gt; GET /lvl-service/rest/v1/sequences HTTP/1.1
&gt; User-Agent: curl/7.35.0
&gt; Host: lvl.i3m.upv.es
&gt; Accept: */*
&gt; 
&lt; HTTP/1.1 200 OK
* Server nginx/1.4.6 (Ubuntu) is not blacklisted
&lt; Server: nginx/1.4.6 (Ubuntu)
&lt; Date: Wed, 17 Sep 2014 10:41:21 GMT
&lt; Content-Type: application/json
&lt; Content-Length: 1586
&lt; Connection: keep-alive
&lt; 
{
    "elements": [
        {
            "accession": "AC238558",
            "countryFeature": null,
            "dataSource": "GenBank",
            "definition": "Phlebotomus papatasi BAC clone PLPAAEX-7H12 from chromosome unknown, complete sequence",
            "gi": 260764140,
            "id": "gb:AC238558",
            "links": [
                {
                    "href": "/lvl-service/rest/v1/sequences/gb:AC238558",
                    "params": {
                        "rel": "self",
                        "type": "application/json"
                    }
                }
            ],
            "locale": null,
            "location": null,
            "organism": "Phlebotomus papatasi",
            "version": "AC238558.1"
        },
        {
            "accession": "AC238775",
            "countryFeature": null,
            "dataSource": "GenBank",
            "definition": "Phlebotomus papatasi BAC clone PLPAAEX-10H12 from chromosome unknown, complete sequence",
            "gi": 262073197,
            "id": "gb:AC238775",
            "links": [
                {
                    "href": "/lvl-service/rest/v1/sequences/gb:AC238775",
                    "params": {
                        "rel": "self",
                        "type": "application/json"
                    }
                }
            ],
            "locale": null,
            "location": null,
            "organism": "Phlebotomus papatasi",
            "version": "AC238775.1"
        },
        {
            "accession": "AC238776",
            "countryFeature": null,
            "dataSource": "GenBank",
            "definition": "Phlebotomus papatasi BAC clone PLPAAEX-12H12 from chromosome unknown, complete sequence",
            "gi": 262073198,
            "id": "gb:AC238776",
            "links": [
                {
                    "href": "/lvl-service/rest/v1/sequences/gb:AC238776",
                    "params": {
                        "rel": "self",
                        "type": "application/json"
                    }
                }
            ],
            "locale": null,
            "location": null,
            "organism": "Phlebotomus papatasi",
            "version": "AC238776.1"
        }
    ],
    "links": [
        {
            "href": "/lvl-service/rest/v1/sequences?page=1&amp;per_page=3&amp;q=&amp;sort=&amp;order=asc",
            "params": {
                "rel": "next",
                "type": "application/json"
            }
        },
        {
            "href": "/lvl-service/rest/v1/sequences?page=8609&amp;per_page=3&amp;q=&amp;sort=&amp;order=asc",
            "params": {
                "rel": "last",
                "type": "application/json"
            }
        }
    ],
    "totalCount": 25828
}
</pre>

            <h3>
              <a id="examples_restful_api_find_nearby_sequences"></a>Find the sequences in the database that are within the specified distance (in meters) from
              the center point specified (using WGS84):
            </h3>

            <p>
              It uses coordinates order longitude, latitude. Response is a GeoJSON Features Collection. There are several free services that you can use to find
              a position on a map. For example the <a href="http://epsg.io/" target="_blank">EPSG.io: Coordinate Systems Worldwide</a> allows using different
              coordinate systems to this end, including <a href="http://epsg.io/4326/map" target="_blank">WGS84</a>.
            </p>

            <p>For example, find sequences located in Yemen (longitude: 48.516388000000006, latitude: 15.552727) within a radius of 2 km:</p>

            <code> $ curl -H 'Authorization: Bearer 4NYglPbKiSx3aNKSY8aIrSnpa3tOj2WytUP+MIaSDB0='
              'http://lvl.i3m.upv.es/lvl-service/rest/v1/sequences/nearby/48.516388000000006/15.552727?maxDistance=2000.0'</code>
            <p></p>

            <pre>
{
    "crs": {
        "properties": {
            "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
        },
        "type": "name"
    },
    "features": [
        {
            "geometry": {
                "coordinates": [
                    48.516388,
                    15.552727
                ],
                "type": "Point"
            },
            "properties": {
                "count": 2,
                "name": "gb:EF408772,gb:EF408798"
            },
            "type": "Feature"
        }
    ],
    "type": "FeatureCollection"
}
</pre>

            <p>To find all the sequences annotated with geospatial coordinates search from the center of the map up to 6,500 km:</p>

            <code>$ curl -H 'Authorization: Bearer 4NYglPbKiSx3aNKSY8aIrSnpa3tOj2WytUP+MIaSDB0='
              'http://lvl.i3m.upv.es/lvl-service/rest/v1/sequences/nearby/0.0/0.0?maxDistance=6500000.0'</code>

            <h3>
              <a id="examples_restful_api_import_sequences"></a>Import new sequences into the database:
            </h3>

            <code> $ curl -v -X POST -H 'Authorization: Bearer 4NYglPbKiSx3aNKSY8aIrSnpa3tOj2WytUP+MIaSDB0=' -H 'Content-Type: application/json' -d '{
              "type" : "IMPORT_SANDFLY_SEQ", "ids" : [ "353470160", "353483325", "384562886" ] }' http://lvl.i3m.upv.es/lvl-service/rest/v1/tasks </code>
            <p></p>

            <pre>
* About to connect() to lvl.i3m.upv.es port 80 (#0)
*   Trying 158.42.105.62... connected
&gt; POST /lvl-service/rest/v1/tasks HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: lvl.i3m.upv.es
&gt; Accept: */*
&gt; Authorization: Bearer 4NYglPbKiSx3aNKSY8aIrSnpa3tOj2WytUP+MIaSDB0=
&gt; Content-Type: application/json
&gt; Content-Length: 82
&gt; 
* upload completely sent off: 82out of 82 bytes
&lt; HTTP/1.1 201 Created
&lt; Server: nginx/1.4.6 (Ubuntu)
&lt; Date: Sat, 10 May 2014 14:31:15 GMT
&lt; Content-Length: 0
&lt; Connection: keep-alive
&lt; Location: http://lvl.i3m.upv.es/lvl-service/rest/v1/tasks/d0464bbc-305a-442c-8b67-34a21fc41af9
&lt; 
* Connection #0 to host lvl.i3m.upv.es left intact
* Closing connection #0
</pre>

            <h3>
              <a id="examples_restful_api_export_public_link_to_sequence"></a>Export a public link to a sequence:
            </h3>
            
            <p>Access to the LVL is provided through a RESTful API requiring authentication to retrieve and to create/modify the data. This behaviour is 
              required to support the requirements identified for the LVL in the project EUBrazilCC. However, accessing from a programmatic environment where 
              users are not available needs an additional mechanism to provide services and applications with anonymous access to the data (read only). The 
              mechanisn implemented in the LVL works in a similar manner as Dropbox links do. Users are able of creating and exporting public links to 
              their data, and any user or application have access to the data provided with a public link.</p>
              
            <p>In addition to these public links, the RESTful API provides additional filters that allow users to export the sequences stored in the LVL 
              database to the most common data formats used in bioinformatics (<a href="http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml" target="_blank">FASTA</a> 
              and <a href="http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html" target="_blank">NCBI</a> are currently supported). Gzip compression 
              is optionally available. The following example shows how a new link to an existing sequence can be created using the RESTful API:</p>
              
<pre>
{
    "description": "Public link to FASTA.GZIP sequences",
    "downloadUri": null,
    "links": null,
    "mime": null,
    "owner": null,
    "path": null,
    "target": {
        "collection": "sandfly",
        "compression": "gzip",
        "filter": "export_fasta",
        "ids": [
            "gb:JP540074",
            "gb:JP553239"
        ],
        "type": "sequence"
    }
}
</pre>              

            <p>The only field required in the request is <code>target</code>. In this case, the origin of the public link is of <code>type</code> sequence
              from the collection of <code>sandfly</code> and two sequence identifiers are passed. A <code>filter</code> is applied to export the 
              sequences to FASTA and GZIP <code>compression</code> is requested. Other possible values are: <code>filter:export</code> to export the 
              sequences in the default format (NCBI XML sequence format), and <code>compression:null</code> or <code>compression:none</code> to export 
              the sequences with no compression.</p>
            
            <div class="alert alert-info">
              <div class="media">
                <div class="pull-left">
                  <h1>
                    <span class="glyphicon glyphicon glyphicon-exclamation-sign"></span>
                  </h1>
                </div>
                <div class="media-body">
                  <h4 class="media-heading">Sequence identifiers</h4>
                  <p>
                    Sequence identifiers include the original database where the sequence was submitted (e.g. GenBank) and the accession number in 
                    the original database (e.g. JP540074). Valid sequence identifiers can be built using both forms the database name: a) long version: 
                    <code>GenBank:JP540074</code>; or b) short version <code>gb:JP540074</code>. Sequence identifiers are case insensitive.
                  </p>
                </div>
              </div>
            </div>
            
            <p>After submitting the request to the LVL, the server response header will include the standar <code>location</code> field.</p>
            
            <code>$ curl -i -H 'Authorization: Bearer a7A1zKUubiomVQdQIuOtdyk8QUHJSpMEl8SE8HrOu0k=' -X POST -H "Content-Type: application/json" -d '{"path":null,"mime":null,"description":"Public link to FASTA.GZIP sequences","owner":null,"target":{"type":"sequence","collection":"sandfly","ids":["gb:JP540074","gb:JP553239"],"filter":"export_fasta","compression":"gzip"},"downloadUri":null,"links":null}' http://lvl.i3m.upv.es/lvl-service/rest/v1/public_links</code>
            <p></p>
            
<pre>
HTTP/1.1 201 Created
Server: nginx/1.4.6 (Ubuntu)
Date: Wed, 24 Sep 2014 11:08:17 GMT
Content-Length: 0
Connection: keep-alive
Location: http://lvl.i3m.upv.es/lvl-service/rest/v1/public_links/bui6vm4cinxg9so5:sequences.fasta.gz
</pre>

            <p>Once the public link is created, the expected operations for consulting, updating and removing it are provided in the RESTful API.</p>

            <p>The public link can be retrieved from the LVL using the location. Note that the server response is always in JSON format, regardless the
              file extension:</p>

            <code>$ curl -H 'Authorization: Bearer a7A1zKUubiomVQdQIuOtdyk8QUHJSpMEl8SE8HrOu0k=' http://lvl.i3m.upv.es/lvl-service/rest/v1/public_links/bui6vm4cinxg9so5:sequences.fasta.gz</code>
            <p></p>

<pre>
{
    "description": "Public link to FASTA.GZIP sequences",
    "downloadUri": "http://lvl.i3m.upv.es/lvl-pub/bui6vm4cinxg9so5/sequences.fasta.gz",
    "links": [
        {
            "href": "/lvl-service/rest/v1/public_links/bui6vm4cinxg9so5:sequences.fasta.gz",
            "params": {
                "rel": "self",
                "type": "application/json"
            }
        }
    ],
    "mime": "application/gzip",
    "owner": "user1",
    "path": "bui6vm4cinxg9so5/sequences.fasta.gz",
    "target": {
        "collection": "sandfly",
        "compression": "gzip",
        "filter": "export_fasta",
        "ids": [
            "gb:JP540074",
            "gb:JP553239"
        ],
        "type": "sequence"
    }
}
</pre>
            
            <p>Additional fields are available in the server response. The <code>downloadUri</code> is especially usefull to download the data from 
              a storage that is optimized for this purpose:</p>
            
            <code>$ wget http://lvl.i3m.upv.es/lvl-pub/bui6vm4cinxg9so5/sequences.fasta.gz</code>
            <p></p>

<pre>
--2014-09-24 13:17:21--  http://lvl.i3m.upv.es/lvl-pub/bui6vm4cinxg9so5/sequences.fasta.gz
Resolving lvl.i3m.upv.es (lvl.i3m.upv.es)... 158.42.105.62
Connecting to lvl.i3m.upv.es (lvl.i3m.upv.es)|158.42.105.62|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 455 [application/octet-stream]
Saving to: ‘sequences.fasta.gz’

100%[======================================>] 455         --.-K/s   in 0s      

2014-09-24 13:17:21 (35.7 MB/s) - ‘sequences.fasta.gz’ saved [455/455]
</pre>

            <p>In this example, the content of the downloaded file is the following:</p>

            <code>$ zmore sequences.fasta.gz</code>
            <p></p>

<pre>
>gb|JP540074.1|JP540074
AAACACTATTCCAGGAAGCCAGACGCATCAACATAGCTATTATTCAACATATTACATACAATGAATTCCT
GCCTATACTCCTGGGCAAGGAAGTGATGGAGAAATTTGGACTGGTTCTACAGAAGGAAGGTTATTGGGAT
GGATATGATGGACACATCAATCCAGCCATCATCGATGCCTTCGCTGCAGCTGCCTTCCGTTTCGGTCATT
CCCTCCTACCAACAGCCGTTGAACGGTGGTCTAAGGCTCACAAATTTATAGCATCGAAGAGGCTGTCCGA
TCTTATCAGACGTCCTTATGATTTGTACAGAGCTGGAGTGTTTGATGAGTATTTCATGGGACTGATGAAT
CAAGTGGCTCAAGCCATGGATGATTCCATCACACAAGAGGTAACAAACCATCTCTTCAAGAAGGAGGGTA
ATCGCTTTGGCATGGACTTGGTGTCCTTCAATATGCAACGAGGAAGAGAATTCGGTTTGCCCGGATATAT
GGAGTTCAAGAAGTTCTGCGGATTGCCAACATCACACACGTGGGAGGAAATGTTCGGTTCGATGCCCAAT
GACACAATTCGACGATACGAGAGTATTTTCCAGCATCCTGCTGACATTGATTTATGGTCTGGTGGAGTTT
ACAAAAGATCACTGCCTGGCTCGATGCTGGGTCCCACTTTCGCTTGC
>gb|JP553239.1|JP553239
AAAATCGCAGGCCGAAATGGGTTTAAAGTACGAACTCTGTGTGGGTCTCAACAGGGGACACAAAACCAGT
AAAATCAGGAATTTGCGGTATACTGGAGACAAGAAGGTGAAAGGACTCCGCCCTGATCGGCTAAAATGTG
TTCAGACGAAACACACAAAATTCGTGAGGGATATTGTGCGTGAGGTGGTTGGACATGCTCCCTATGAGAA
GCGCTGCATGGAATTGCTGAAGGTGTCTAAGGATAAGCGTGCCCTGAAATTCCTGAAACGTCGTCTGGGC
ACTCACATCCGTGGCAAGAGGAAGCGTGAGGAACTGTCCAACATTCTGACACAAATGCGCAAGGCTCAGG
CCCATGCCAAGTAAAATTTTTCCAGGAGGAATAAAGATGATT
</pre>

            <p>Using NCBI format, the output is as follows:</p>
            
<pre>
&lt;?xml version="1.0" encoding="UTF-8" standalone="yes"?&gt;
&lt;GBSet&gt;
  &lt;GBSeq&gt;
    &lt;GBSeq_locus&gt;JP540074&lt;/GBSeq_locus&gt;
    &lt;GBSeq_length&gt;677&lt;/GBSeq_length&gt;
    &lt;GBSeq_strandedness&gt;single&lt;/GBSeq_strandedness&gt;
    &lt;GBSeq_moltype&gt;mRNA&lt;/GBSeq_moltype&gt;
    &lt;GBSeq_topology&gt;linear&lt;/GBSeq_topology&gt;
    &lt;GBSeq_division&gt;TSA&lt;/GBSeq_division&gt;
    &lt;GBSeq_update-date&gt;25-MAR-2013&lt;/GBSeq_update-date&gt;
    &lt;GBSeq_create-date&gt;14-OCT-2012&lt;/GBSeq_create-date&gt;
    &lt;GBSeq_definition&gt;TSA: Phlebotomus papatasi PRAG-aaa24b11.b1 mRNA sequence&lt;/GBSeq_definition&gt;
    &lt;GBSeq_primary-accession&gt;JP540074&lt;/GBSeq_primary-accession&gt;
    &lt;GBSeq_accession-version&gt;JP540074.1&lt;/GBSeq_accession-version&gt;
    &lt;GBSeq_other-seqids&gt;
      &lt;GBSeqid&gt;gnl|gpid:68651|PRAG-aaa24b11.b1&lt;/GBSeqid&gt;
      &lt;GBSeqid&gt;gb|JP540074.1|&lt;/GBSeqid&gt;
      &lt;GBSeqid&gt;gi|353470160&lt;/GBSeqid&gt;
    &lt;/GBSeq_other-seqids&gt;
    &lt;GBSeq_project&gt;PRJNA68651&lt;/GBSeq_project&gt;
    &lt;GBSeq_keywords&gt;
      &lt;GBKeyword&gt;TSA&lt;/GBKeyword&gt;
      &lt;GBKeyword&gt;Transcriptome Shotgun Assembly&lt;/GBKeyword&gt;
    &lt;/GBSeq_keywords&gt;
    &lt;GBSeq_source&gt;Phlebotomus papatasi&lt;/GBSeq_source&gt;
    &lt;GBSeq_organism&gt;Phlebotomus papatasi&lt;/GBSeq_organism&gt;
    &lt;GBSeq_taxonomy&gt;Eukaryota; Metazoa; Ecdysozoa; Arthropoda; Hexapoda; Insecta; Pterygota; Neoptera; Endopterygota; Diptera; Nematocera; Psychodoidea; Psychodidae; Phlebotomus; Phlebotomus&lt;/GBSeq_taxonomy&gt;
    &lt;GBSeq_references&gt;
      &lt;GBReference&gt;
        &lt;GBReference_reference&gt;1&lt;/GBReference_reference&gt;
        &lt;GBReference_position&gt;1..677&lt;/GBReference_position&gt;
        &lt;GBReference_authors&gt;
          &lt;GBAuthor&gt;Abrudan,J.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Ramalho-Ortigao,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;O'Neil,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Stayback,G.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Wadsworth,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Bernard,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Shoue,D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Emrich,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lawyer,P.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Kamhawi,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Rowton,E.D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lehane,M.J.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Bates,P.A.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Valenzeula,J.G.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Tomlinson,C.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Appelbaum,E.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Moeller,D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Thiesing,B.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Dillon,R.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Clifton,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lobo,N.F.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Wilson,R.K.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Collins,F.H.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;McDowell,M.A.&lt;/GBAuthor&gt;
        &lt;/GBReference_authors&gt;
        &lt;GBReference_title&gt;The characterization of the Phlebotomus papatasi transcriptome&lt;/GBReference_title&gt;
        &lt;GBReference_journal&gt;Insect Mol. Biol. 22 (2), 211-232 (2013)&lt;/GBReference_journal&gt;
        &lt;GBReference_xref&gt;
          &lt;GBXref&gt;
            &lt;GBXref_dbname&gt;doi&lt;/GBXref_dbname&gt;
            &lt;GBXref_id&gt;10.1111/imb.12015&lt;/GBXref_id&gt;
          &lt;/GBXref&gt;
        &lt;/GBReference_xref&gt;
        &lt;GBReference_pubmed&gt;23398403&lt;/GBReference_pubmed&gt;
      &lt;/GBReference&gt;
      &lt;GBReference&gt;
        &lt;GBReference_reference&gt;2&lt;/GBReference_reference&gt;
        &lt;GBReference_position&gt;1..677&lt;/GBReference_position&gt;
        &lt;GBReference_authors&gt;
          &lt;GBAuthor&gt;Abrudan,J.L.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;McDowell,M.A.&lt;/GBAuthor&gt;
        &lt;/GBReference_authors&gt;
        &lt;GBReference_title&gt;Direct Submission&lt;/GBReference_title&gt;
        &lt;GBReference_journal&gt;Submitted (11-OCT-2011) Biology, University of Notre Dame, 215 Galvin, Notre Dame, IN 46556, USA&lt;/GBReference_journal&gt;
      &lt;/GBReference&gt;
    &lt;/GBSeq_references&gt;
    &lt;GBSeq_comment&gt;##Assembly-Data-START## Assembly Method Cap3 Sequencing Technology Sanger ##Assembly-Data-END##&lt;/GBSeq_comment&gt;
    &lt;GBSeq_feature-table&gt;
      &lt;GBFeature&gt;
        &lt;GBFeature_key&gt;source&lt;/GBFeature_key&gt;
        &lt;GBFeature_location&gt;1..677&lt;/GBFeature_location&gt;
        &lt;GBFeature_intervals&gt;
          &lt;GBInterval&gt;
            &lt;GBInterval_from&gt;1&lt;/GBInterval_from&gt;
            &lt;GBInterval_to&gt;677&lt;/GBInterval_to&gt;
            &lt;GBInterval_accession&gt;JP540074.1&lt;/GBInterval_accession&gt;
          &lt;/GBInterval&gt;
        &lt;/GBFeature_intervals&gt;
        &lt;GBFeature_quals&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;organism&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;Phlebotomus papatasi&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;mol_type&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;mRNA&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;strain&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;Israeli&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;db_xref&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;taxon:29031&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
        &lt;/GBFeature_quals&gt;
      &lt;/GBFeature&gt;
    &lt;/GBSeq_feature-table&gt;
    &lt;GBSeq_sequence&gt;aaacactattccaggaagccagacgcatcaacatagctattattcaacatattacatacaatgaattcctgcctatactcctgggcaaggaagtgatggagaaatttggactggttctacagaaggaaggttattgggatggatatgatggacacatcaatccagccatcatcgatgccttcgctgcagctgccttccgtttcggtcattccctcctaccaacagccgttgaacggtggtctaaggctcacaaatttatagcatcgaagaggctgtccgatcttatcagacgtccttatgatttgtacagagctggagtgtttgatgagtatttcatgggactgatgaatcaagtggctcaagccatggatgattccatcacacaagaggtaacaaaccatctcttcaagaaggagggtaatcgctttggcatggacttggtgtccttcaatatgcaacgaggaagagaattcggtttgcccggatatatggagttcaagaagttctgcggattgccaacatcacacacgtgggaggaaatgttcggttcgatgcccaatgacacaattcgacgatacgagagtattttccagcatcctgctgacattgatttatggtctggtggagtttacaaaagatcactgcctggctcgatgctgggtcccactttcgcttgc&lt;/GBSeq_sequence&gt;
    &lt;GBSeq_xrefs&gt;
      &lt;GBXref&gt;
        &lt;GBXref_dbname&gt;BioProject&lt;/GBXref_dbname&gt;
        &lt;GBXref_id&gt;PRJNA68651&lt;/GBXref_id&gt;
      &lt;/GBXref&gt;
    &lt;/GBSeq_xrefs&gt;
  &lt;/GBSeq&gt;
  &lt;GBSeq&gt;
    &lt;GBSeq_locus&gt;JP553239&lt;/GBSeq_locus&gt;
    &lt;GBSeq_length&gt;392&lt;/GBSeq_length&gt;
    &lt;GBSeq_strandedness&gt;single&lt;/GBSeq_strandedness&gt;
    &lt;GBSeq_moltype&gt;mRNA&lt;/GBSeq_moltype&gt;
    &lt;GBSeq_topology&gt;linear&lt;/GBSeq_topology&gt;
    &lt;GBSeq_division&gt;TSA&lt;/GBSeq_division&gt;
    &lt;GBSeq_update-date&gt;25-MAR-2013&lt;/GBSeq_update-date&gt;
    &lt;GBSeq_create-date&gt;14-OCT-2012&lt;/GBSeq_create-date&gt;
    &lt;GBSeq_definition&gt;TSA: Phlebotomus papatasi PpContig1430.f1 mRNA sequence&lt;/GBSeq_definition&gt;
    &lt;GBSeq_primary-accession&gt;JP553239&lt;/GBSeq_primary-accession&gt;
    &lt;GBSeq_accession-version&gt;JP553239.1&lt;/GBSeq_accession-version&gt;
    &lt;GBSeq_other-seqids&gt;
      &lt;GBSeqid&gt;gnl|gpid:68651|PpContig1430.f1&lt;/GBSeqid&gt;
      &lt;GBSeqid&gt;gb|JP553239.1|&lt;/GBSeqid&gt;
      &lt;GBSeqid&gt;gi|353483325&lt;/GBSeqid&gt;
    &lt;/GBSeq_other-seqids&gt;
    &lt;GBSeq_project&gt;PRJNA68651&lt;/GBSeq_project&gt;
    &lt;GBSeq_keywords&gt;
      &lt;GBKeyword&gt;TSA&lt;/GBKeyword&gt;
      &lt;GBKeyword&gt;Transcriptome Shotgun Assembly&lt;/GBKeyword&gt;
    &lt;/GBSeq_keywords&gt;
    &lt;GBSeq_source&gt;Phlebotomus papatasi&lt;/GBSeq_source&gt;
    &lt;GBSeq_organism&gt;Phlebotomus papatasi&lt;/GBSeq_organism&gt;
    &lt;GBSeq_taxonomy&gt;Eukaryota; Metazoa; Ecdysozoa; Arthropoda; Hexapoda; Insecta; Pterygota; Neoptera; Endopterygota; Diptera; Nematocera; Psychodoidea; Psychodidae; Phlebotomus; Phlebotomus&lt;/GBSeq_taxonomy&gt;
    &lt;GBSeq_references&gt;
      &lt;GBReference&gt;
        &lt;GBReference_reference&gt;1&lt;/GBReference_reference&gt;
        &lt;GBReference_position&gt;1..392&lt;/GBReference_position&gt;
        &lt;GBReference_authors&gt;
          &lt;GBAuthor&gt;Abrudan,J.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Ramalho-Ortigao,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;O'Neil,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Stayback,G.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Wadsworth,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Bernard,M.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Shoue,D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Emrich,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lawyer,P.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Kamhawi,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Rowton,E.D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lehane,M.J.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Bates,P.A.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Valenzeula,J.G.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Tomlinson,C.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Appelbaum,E.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Moeller,D.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Thiesing,B.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Dillon,R.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Clifton,S.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Lobo,N.F.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Wilson,R.K.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;Collins,F.H.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;McDowell,M.A.&lt;/GBAuthor&gt;
        &lt;/GBReference_authors&gt;
        &lt;GBReference_title&gt;The characterization of the Phlebotomus papatasi transcriptome&lt;/GBReference_title&gt;
        &lt;GBReference_journal&gt;Insect Mol. Biol. 22 (2), 211-232 (2013)&lt;/GBReference_journal&gt;
        &lt;GBReference_xref&gt;
          &lt;GBXref&gt;
            &lt;GBXref_dbname&gt;doi&lt;/GBXref_dbname&gt;
            &lt;GBXref_id&gt;10.1111/imb.12015&lt;/GBXref_id&gt;
          &lt;/GBXref&gt;
        &lt;/GBReference_xref&gt;
        &lt;GBReference_pubmed&gt;23398403&lt;/GBReference_pubmed&gt;
      &lt;/GBReference&gt;
      &lt;GBReference&gt;
        &lt;GBReference_reference&gt;2&lt;/GBReference_reference&gt;
        &lt;GBReference_position&gt;1..392&lt;/GBReference_position&gt;
        &lt;GBReference_authors&gt;
          &lt;GBAuthor&gt;Abrudan,J.L.&lt;/GBAuthor&gt;
          &lt;GBAuthor&gt;McDowell,M.A.&lt;/GBAuthor&gt;
        &lt;/GBReference_authors&gt;
        &lt;GBReference_title&gt;Direct Submission&lt;/GBReference_title&gt;
        &lt;GBReference_journal&gt;Submitted (11-OCT-2011) Biology, University of Notre Dame, 215 Galvin, Notre Dame, IN 46556, USA&lt;/GBReference_journal&gt;
      &lt;/GBReference&gt;
    &lt;/GBSeq_references&gt;
    &lt;GBSeq_comment&gt;##Assembly-Data-START## Assembly Method Cap3 Sequencing Technology Sanger ##Assembly-Data-END##&lt;/GBSeq_comment&gt;
    &lt;GBSeq_feature-table&gt;
      &lt;GBFeature&gt;
        &lt;GBFeature_key&gt;source&lt;/GBFeature_key&gt;
        &lt;GBFeature_location&gt;1..392&lt;/GBFeature_location&gt;
        &lt;GBFeature_intervals&gt;
          &lt;GBInterval&gt;
            &lt;GBInterval_from&gt;1&lt;/GBInterval_from&gt;
            &lt;GBInterval_to&gt;392&lt;/GBInterval_to&gt;
            &lt;GBInterval_accession&gt;JP553239.1&lt;/GBInterval_accession&gt;
          &lt;/GBInterval&gt;
        &lt;/GBFeature_intervals&gt;
        &lt;GBFeature_quals&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;organism&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;Phlebotomus papatasi&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;mol_type&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;mRNA&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;strain&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;Israeli&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
          &lt;GBQualifier&gt;
            &lt;GBQualifier_name&gt;db_xref&lt;/GBQualifier_name&gt;
            &lt;GBQualifier_value&gt;taxon:29031&lt;/GBQualifier_value&gt;
          &lt;/GBQualifier&gt;
        &lt;/GBFeature_quals&gt;
      &lt;/GBFeature&gt;
    &lt;/GBSeq_feature-table&gt;
    &lt;GBSeq_sequence&gt;aaaatcgcaggccgaaatgggtttaaagtacgaactctgtgtgggtctcaacaggggacacaaaaccagtaaaatcaggaatttgcggtatactggagacaagaaggtgaaaggactccgccctgatcggctaaaatgtgttcagacgaaacacacaaaattcgtgagggatattgtgcgtgaggtggttggacatgctccctatgagaagcgctgcatggaattgctgaaggtgtctaaggataagcgtgccctgaaattcctgaaacgtcgtctgggcactcacatccgtggcaagaggaagcgtgaggaactgtccaacattctgacacaaatgcgcaaggctcaggcccatgccaagtaaaatttttccaggaggaataaagatgatt&lt;/GBSeq_sequence&gt;
    &lt;GBSeq_xrefs&gt;
      &lt;GBXref&gt;
        &lt;GBXref_dbname&gt;BioProject&lt;/GBXref_dbname&gt;
        &lt;GBXref_id&gt;PRJNA68651&lt;/GBXref_id&gt;
      &lt;/GBXref&gt;
    &lt;/GBSeq_xrefs&gt;
  &lt;/GBSeq&gt;
&lt;/GBSet&gt;
</pre>

            <h2>
              <a id="examples_oauth2"></a>OAuth2 examples
            </h2>

            <p>
              This section provides simple examples to illustrate the use of the OAuth2 API. The examples use <a href="http://curl.haxx.se/" target="_blank">cURL</a>
              to access the endpoint.
            </p>

            <h3>
              <a id="examples_oauth2_get_access_token_password"></a>Get access token for user/password:
            </h3>

            <code> $ curl -v -X POST -d 'client_id=lvl_portal&amp;client_secret=changeit&amp;grant_type=password&amp;username=root&amp;password=changeit'
              http://lvl.i3m.upv.es/lvl-auth/oauth2/v1/token </code>
            <p></p>

            <pre>
* About to connect() to lvl.i3m.upv.es port 80 (#0)
*   Trying 158.42.105.62... connected
&gt; POST /lvl-auth/oauth2/v1/token HTTP/1.1
&gt; User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
&gt; Host: lvl.i3m.upv.es
&gt; Accept: */*
&gt; Content-Length: 95
&gt; Content-Type: application/x-www-form-urlencoded
&gt; 
* upload completely sent off: 95out of 95 bytes
&lt; HTTP/1.1 200 OK
&lt; Server: nginx/1.1.19
&lt; Date: Tue, 15 Apr 2014 10:19:36 GMT
&lt; Content-Type: application/json
&lt; Content-Length: 164
&lt; Connection: keep-alive
&lt; Access-Control-Allow-Origin: *
&lt; Access-Control-Allow-Headers: origin, content-type, accept, authorization
&lt; Access-Control-Allow-Credentials: true
&lt; Access-Control-Allow-Methods: GET PUT DELETE POST HEAD OPTIONS
&lt; Access-Control-Max-Age: 604800
&lt; 
* Connection #0 to host lvl.i3m.upv.es left intact
* Closing connection #0
{"scope":"users/*,a sequences,a pipelines,a my_sequences/*,a pending_sequences,a","expires_in":604800,"access_token":"+sSWJ1WLssx00iT6obLw8jxPNlCtvJCN9a17Twwqnuo="}
</pre>

            <h3>
              <a id="examples_oauth2_access_to_protected_resource"></a>Access to protected resource:
            </h3>

            <code> $ curl -H "Authorization: Bearer k94LAp+oPa7bMNVCut84II7VTR+vEpf6MW+GaQHK72U="
              http://lvl.i3m.upv.es/lvl-auth/oauth2/v1/users/root@lvl </code>
            <p></p>

<pre>
{
    "email": "***",
    "fullname": "LVL root user",
    "links": [
        {
            "href": "/lvl-auth/oauth2/v1/users/root",
            "params": {
                "rel": "self",
                "type": "application/json"
            }
        }
    ],
    "password": "***",
    "permissionHistory": {
        "history": {}
    },
    "permissions": [
        "*"
    ],
    "pictureUrl": "https://secure.gravatar.com/avatar/04edfc0ef6c6cf6d6b88fbc69f9f9071?s=80&amp;d=mm",
    "provider": "lvl",
    "roles": [
        "admin"
    ],
    "salt": "***",
    "userid": "root"

</pre>

            <h2>
              <a id="examples_tasks"></a>Task examples
            </h2>

            <p>
              This section provides simple examples to illustrate the use of the REST API to execute tasks in the LVL. The examples use 
              <a href="http://curl.haxx.se/" target="_blank">cURL</a> to access the endpoint.
            </p>

            <h3>
              <a id="examples_tasks_import_sequences"></a>Import sequences from external databases (PubMed):
            </h3>

            <code> $ curl -H 'Authorization: Bearer ubauFeNkTWiVGQWTcfJPUW55nWTIY3c+yu25p/Sm7KU=' -X POST -H "Content-Type: application/json" 
              -d '{"uuid":null,"type":"IMPORT_SANDFLY_SEQ"}' http://lvl.i3m.upv.es/lvl-service/rest/v1/tasks </code>
            <p></p>

          </div>
        </div>
      </div>
    </div>
    
    <div class="clearfix">
      <div class="row">
        <div class="col-xs-12">
          <div class="page-header">
            <h1 id="type">
              <a id="limitations"></a>Known limitations
            </h1>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">
            <h2>
              <a id="limitations_usage_limits"></a>Usage limits
            </h2>

            <p>The Leishmaniasis Virtual Laboratory is a non-profit project mainly funded with public means. Therefore, many services have usage limitations
              with the objective of providing the best possible service to the most users. The following list describe the current limitations of the LVL service:
            </p>
            
            <ul>
              <li>The Google Geocoding API imposes <a href="https://developers.google.com/maps/documentation/geocoding/#Limits" target="_blank">usage limits</a>
                to free users. When these limits are overpassed the error <code>OVER_QUERY_LIMIT</code> is returned to the caller.</li>
            </ul>
                                    
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-xs-12">
          <div class="bs-component">
            <h2>
              <a id="limitations_desired_features"></a>Desired features list
            </h2>
            
            <ul>
              <li>ActiveMQ proxied? - http://activemq.apache.org/http-and-https-transports-reference.html</li>
              <li>Store tasks with the LVL database:
                <ul>
                  <li>Tasks must be stored in the database of the LVL and the service replica that runs the task should update the status in the database, 
                    while the other service replicas should read the status of the task from the database.</li></ul></li>
              <li>Extend user support to include social network capabilities.
                <ul><li>Add the following fields to the current user specification: location, country, joining date, description (optional). Create lists of
                  followers/following.</li></ul></li>
              <li>Implement periodic maintenance tasks:
                <ul>
                  <li>Periodically remove from the LVL database tasks completed after 30 days.</li>
                  <li>Periodically fetch new sequences from external databases.</li>
                  <li>An option could be to integrate the <a href="http://www.javacodegeeks.com/2012/04/configuring-quartz-with-jdbcjobstore-in.html" target="_blank">Quartz</a>
                    scheduler into the LVL service with a permanent store.</li>
                </ul>
              </li>
              <li>Improve user experience:
                <ul>
                  <li>Migrate login waiting sreeen by the <a href="http://www.webcodegeeks.com/javascript/boot-your-ajax-app-creating-a-splash-screen-with-nprogress/" target="_blank">NProgress</a>
                    library.</li>
                  <li>Communicate between views: <a href="http://www.webcodegeeks.com/javascript/backbone-js/communicating-between-views-in-backbone/" target="_blank">here</a>.</li>
                </ul>
              </li>
              <li>Improve the inference of geographic coordinates from DNA sequences imported from GenBank.</li>
              <li>Improve integration with Apache Shiro: Errors are logged when the LVL application is stopped/undeployed in Tomcat due to an unclean 
              shutdown of Apache Shiro. These are not critical, but produces large error messages.
                <ul>
                  <li>Reference: <a href="http://mail-archives.apache.org/mod_mbox/shiro-user/201210.mbox/%3CCAAtvD4XVNhKt=2m5yJecKOYxr+PJDraMOWqoodakjk1wWxkh-g@mail.gmail.com%3E">Unclean shutdown of Tomcat</a>.</li>
                </ul>
              </li>
            </ul>
            
            <p>
              Currently, the geographic coordinates are inferred from the country feature of the GenBank record using the Google Geocoding API. However, this
              field is annotated with a free-text format, difficulting the inference in many cases. When this occurs a <code>ZERO_RESULTS</code> error is
              returned to the caller. The following examples were extracted from the logging system of the LVL service, demonstrating that in many cases
              there are misspelling errors in the records or mixture of terms expressed in different languages:
            </p>
            
            <code>$ sed '/ZERO_RESULTS/!d;s/^java.lang.IllegalStateException: Searching for //;s/ produces invalid Google Geocoding server response: ZERO_RESULTS$//' lvl-service.log | sort -u</code>
            <p></p>
            
<pre>            
'Algeria: Adama'
'Algeria: Fdoules'
'Algeria: Khola'
'Algeria: Madjana'
'Algeria: Vieux Mila'
'Brazil: Keur Moussa, Senegal'
'Cyprus: Mandria cave'
'Egypt: colony from Southern Sina'
'Egypt: Sinai, colony'
'France: Rosis, Cevennes area'
'France: Trois-puits, near Heims'
'Greece: Psili Rahi'
'Kenya: Utut Reserve'
'Lebanon: Drya'
'Lebanon: Nachgura'
'Madagascar: Androdroka'
'Madagascar: Bemaraha-Anjohikinakina'
'Madagascar: Mahajanga-Anjohikely'
'Mali: Sougoula village'
'Montenegro: Dabcevici'
'Oman: Djebel Quara'
'Oman: Rowda'
'Saudi Arabia: Hofouf, colony'
'Sudan: Moyleih, Khartoum State'
'Sudan: Sirougia, Khartoum State'
'Sudan: Wad Shariefai, Kassala State'
'Syria: Rhuaibe'
'Syria: Rhuhaibe'
'Thailand: Ratchabun province, Muang district, Huay-Phai sub-district'
</pre>      

          </div>
        </div>
      </div>
    </div>

    <hr />

    <footer>
      <div class="row">
        <div class="col-xs-12">
          <ul class="list-unstyled">
            <li class="pull-right"><a href="#top">Back to top</a></li>
          </ul>
          <h4>Leishmaniasis Virtual Laboratory</h4>
          <p>
            Copyright &copy; 2014 <a href="http://www.eubrazilcloudconnect.eu/" target="_blank">EUBrazilCC</a>, Licensed under the <a
              href="https://joinup.ec.europa.eu/software/page/eupl" target="_blank">European Union Public Licence (EUPL)</a>.
          </p>
        </div>
      </div>
    </footer>

    <h5>&nbsp;</h5>

  </div>

  <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
  <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
</body>
</html>